// antdv-next@1.2.2/dist/index.js
import version_default from "./version/index.js";
import UniqueProvider_default from "./tooltip/UniqueProvider/index.js";
import tooltip_default from "./tooltip/index.js";
import { useExportConfig } from "./config-provider/hooks/useConfig.js";
import config_provider_default from "./config-provider/index.js";
import affix_default from "./affix/index.js";
import alert_default from "./alert/index.js";
import anchor_default from "./anchor/index.js";
import useMessage from "./message/useMessage.js";
import message_default from "./message/index.js";
import CollapsePanel_default from "./collapse/CollapsePanel.js";
import collapse_default from "./collapse/index.js";
import button_default from "./button/index.js";
import Avatar_default from "./skeleton/Avatar.js";
import Button_default from "./skeleton/Button.js";
import Node_default from "./skeleton/Node.js";
import Image_default from "./skeleton/Image.js";
import Input_default from "./skeleton/Input.js";
import skeleton_default from "./skeleton/index.js";
import useModal from "./modal/useModal/index.js";
import modal_default from "./modal/index.js";
import useNotification from "./notification/useNotification.js";
import notification_default from "./notification/index.js";
import app_default from "./app/index.js";
import empty_default from "./empty/index.js";
import select_default, { SelectOptGroup, SelectOption } from "./select/index.js";
import auto_complete_default, { Option } from "./auto-complete/index.js";
import popover_default from "./popover/index.js";
import AvatarGroup_default from "./avatar/AvatarGroup.js";
import avatar_default from "./avatar/index.js";
import badge_default, { BadgeRibbon } from "./badge/index.js";
import MenuDivider_default from "./menu/MenuDivider.js";
import SubMenu_default from "./menu/SubMenu.js";
import MenuItemGroup_default from "./menu/MenuItemGroup.js";
import menu_default, { MenuItem } from "./menu/index.js";
import dropdown_default from "./dropdown/index.js";
import BreadcrumbSeparator_default from "./breadcrumb/BreadcrumbSeparator.js";
import BreadcrumbItem_default from "./breadcrumb/BreadcrumbItem.js";
import breadcrumb_default from "./breadcrumb/index.js";
import radio_default, { RadioButton, RadioGroup } from "./radio/index.js";
import calendar_default from "./calendar/index.js";
import TabPane_default from "./tabs/TabPane.js";
import tabs_default from "./tabs/index.js";
import CardGrid_default from "./card/CardGrid.js";
import CardMeta_default from "./card/CardMeta.js";
import card_default from "./card/index.js";
import carousel_default from "./carousel/index.js";
import Panel_default from "./cascader/Panel.js";
import cascader_default from "./cascader/index.js";
import Group_default from "./checkbox/Group.js";
import checkbox_default from "./checkbox/index.js";
import divider_default from "./divider/index.js";
import segmented_default from "./segmented/index.js";
import space_default, { SpaceAddon, SpaceCompact } from "./space/index.js";
import input_number_default from "./input-number/index.js";
import slider_default from "./slider/index.js";
import color_picker_default from "./color-picker/index.js";
import date_picker_default, { DateMonthPicker, DateQuarterPicker, DateRangePicker, DateWeekPicker, DateYearPicker } from "./date-picker/index.js";
import col_default from "./grid/col.js";
import row_default from "./grid/row.js";
import { useBreakpoint } from "./grid/index.js";
import Item_default from "./descriptions/Item.js";
import descriptions_default from "./descriptions/index.js";
import drawer_default from "./drawer/index.js";
import flex_default from "./flex/index.js";
import BackTop_default from "./float-button/BackTop.js";
import FloatButtonGroup_default from "./float-button/FloatButtonGroup.js";
import float_button_default from "./float-button/index.js";
import FormItem_default from "./form/FormItem/index.js";
import form_default from "./form/index.js";
import image_default, { ImagePreviewGroup } from "./image/index.js";
import TextArea_default from "./input/TextArea.js";
import input_default, { InputGroup, InputOTP, InputPassword, InputSearch } from "./input/index.js";
import layout_default, { LayoutContent, LayoutFooter, LayoutHeader, LayoutSider, useLayoutSider } from "./layout/index.js";
import masonry_default from "./masonry/index.js";
import spin_default from "./spin/index.js";
import mentions_default, { Option as Option$1 } from "./mentions/index.js";
import pagination_default from "./pagination/index.js";
import popconfirm_default from "./popconfirm/index.js";
import progress_default from "./progress/progress.js";
import qrcode_default from "./qrcode/index.js";
import rate_default from "./rate/index.js";
import result_default from "./result/index.js";
import splitter_default, { SplitterPanel } from "./splitter/index.js";
import statistic_default, { StatisticTimer } from "./statistic/index.js";
import steps_default from "./steps/index.js";
import switch_default from "./switch/index.js";
import Column_default from "./table/Column.js";
import ColumnGroup_default from "./table/ColumnGroup.js";
import DirectoryTree_default from "./tree/DirectoryTree.js";
import tree_default from "./tree/index.js";
import table_default, { Summary, SummaryCell, SummaryRow } from "./table/index.js";
import CheckableTag_default from "./tag/CheckableTag.js";
import CheckableTagGroup_default from "./tag/CheckableTagGroup.js";
import tag_default from "./tag/index.js";
import time_picker_default, { TimeRangePicker } from "./time-picker/index.js";
import { TimelineItem } from "./timeline/TimelineItem.js";
import timeline_default from "./timeline/index.js";
import tour_default from "./tour/index.js";
import transfer_default from "./transfer/index.js";
import tree_select_default from "./tree-select/index.js";
import typography_default, { TypographyLink, TypographyParagraph, TypographyText, TypographyTitle } from "./typography/index.js";
import upload_default, { UploadDragger } from "./upload/index.js";
import watermark_default from "./watermark/index.js";
import { components_exports } from "./components.js";
import { useResponsive } from "./_util/hooks/useResponsive.js";
import theme_default from "./theme/index.js";
import { StyleProvider } from "@antdv-next/cssinjs";

//#region src/index.ts
let prefix = "A";
var src_default = {
	setPrefix(newPrefix) {
		prefix = newPrefix;
	},
	install(app) {
		app.config.globalProperties._ant_prefix = prefix;
		Object.keys(components_exports).forEach((key) => {
			const component = components_exports[key];
			if ("install" in component) app.use(component);
		});
		app.component("AStyleProvider", StyleProvider);
	},
	version: version_default
};

//#endregion
export { affix_default as Affix, alert_default as Alert, anchor_default as Anchor, app_default as App, auto_complete_default as AutoComplete, Option as AutoCompleteOption, avatar_default as Avatar, AvatarGroup_default as AvatarGroup, BackTop_default as BackTop, badge_default as Badge, BadgeRibbon, breadcrumb_default as Breadcrumb, BreadcrumbItem_default as BreadcrumbItem, BreadcrumbSeparator_default as BreadcrumbSeparator, button_default as Button, calendar_default as Calendar, card_default as Card, CardGrid_default as CardGrid, CardMeta_default as CardMeta, carousel_default as Carousel, cascader_default as Cascader, Panel_default as CascaderPanel, CheckableTag_default as CheckableTag, CheckableTagGroup_default as CheckableTagGroup, checkbox_default as Checkbox, Group_default as CheckboxGroup, col_default as Col, collapse_default as Collapse, CollapsePanel_default as CollapsePanel, color_picker_default as ColorPicker, config_provider_default as ConfigProvider, DateMonthPicker, date_picker_default as DatePicker, DateQuarterPicker, DateRangePicker, DateWeekPicker, DateYearPicker, descriptions_default as Descriptions, Item_default as DescriptionsItem, DirectoryTree_default as DirectoryTree, divider_default as Divider, drawer_default as Drawer, dropdown_default as Dropdown, empty_default as Empty, flex_default as Flex, float_button_default as FloatButton, FloatButtonGroup_default as FloatButtonGroup, form_default as Form, FormItem_default as FormItem, image_default as Image, ImagePreviewGroup, input_default as Input, InputGroup, input_number_default as InputNumber, InputOTP, InputPassword, InputSearch, layout_default as Layout, LayoutContent, LayoutFooter, LayoutHeader, LayoutSider, masonry_default as Masonry, mentions_default as Mentions, Option$1 as MentionsOption, menu_default as Menu, MenuDivider_default as MenuDivider, MenuItem, MenuItemGroup_default as MenuItemGroup, modal_default as Modal, pagination_default as Pagination, popconfirm_default as Popconfirm, popover_default as Popover, progress_default as Progress, qrcode_default as QRCode, radio_default as Radio, RadioButton, RadioGroup, rate_default as Rate, result_default as Result, row_default as Row, segmented_default as Segmented, select_default as Select, SelectOptGroup, SelectOption, skeleton_default as Skeleton, Avatar_default as SkeletonAvatar, Button_default as SkeletonButton, Image_default as SkeletonImage, Input_default as SkeletonInput, Node_default as SkeletonNode, slider_default as Slider, space_default as Space, SpaceAddon, SpaceCompact, spin_default as Spin, splitter_default as Splitter, SplitterPanel, statistic_default as Statistic, StatisticTimer, steps_default as Steps, StyleProvider, SubMenu_default as SubMenu, switch_default as Switch, TabPane_default as TabPane, table_default as Table, Column_default as TableColumn, ColumnGroup_default as TableColumnGroup, Summary as TableSummary, SummaryCell as TableSummaryCell, SummaryRow as TableSummaryRow, tabs_default as Tabs, tag_default as Tag, TextArea_default as TextArea, time_picker_default as TimePicker, TimeRangePicker, timeline_default as Timeline, TimelineItem, tooltip_default as Tooltip, tour_default as Tour, transfer_default as Transfer, tree_default as Tree, tree_select_default as TreeSelect, typography_default as Typography, TypographyLink, TypographyParagraph, TypographyText, TypographyTitle, UniqueProvider_default as UniqueProvider, upload_default as Upload, UploadDragger, watermark_default as Watermark, src_default as default, message_default as message, notification_default as notification, theme_default as theme, useBreakpoint, useExportConfig as useConfig, useLayoutSider, useMessage, useModal, useNotification, useResponsive, version_default as version };
