// @v-c/menu@1.0.14/dist/SubMenu/InlineSubMenuList.js
import InheritableContextProvider, { useMenuContext } from "../context/MenuContext.js";
import { getMotion } from "../utils/motionUtil.js";
import InternalSubMenuList from "./SubMenuList.js";
import { Transition, computed, createVNode, defineComponent, shallowRef, vShow, watch, watchEffect, withDirectives } from "vue";
import { getTransitionProps } from "@v-c/util/dist/utils/transition";
//#region src/SubMenu/InlineSubMenuList.tsx
var InlineSubMenuList = /* @__PURE__ */ defineComponent((props, { slots }) => {
	const fixedMode = "inline";
	const menuContext = useMenuContext();
	const sameModeRef = shallowRef(false);
	watchEffect(() => {
		sameModeRef.value = menuContext?.value?.mode === fixedMode;
	});
	const destroy = shallowRef(!sameModeRef.value);
	watch(() => menuContext?.value?.mode, () => {
		if (sameModeRef.value) destroy.value = false;
	}, { immediate: true });
	const mergedOpen = computed(() => sameModeRef.value ? props?.open : false);
	const mergedMotion = computed(() => {
		const { motion, defaultMotions } = menuContext?.value ?? {};
		const motionData = { ...getMotion(fixedMode, motion, defaultMotions) };
		if (props.keyPath && props.keyPath.length > 1) motionData.appear = false;
		const _onAfterLeave = motionData.onAfterLeave;
		motionData.onAfterLeave = (el) => {
			if (!sameModeRef.value) destroy.value = true;
			return _onAfterLeave?.(el);
		};
		return motionData;
	});
	return () => {
		if (destroy.value) return null;
		return createVNode(InheritableContextProvider, {
			"mode": fixedMode,
			"locked": !sameModeRef.value
		}, { default: () => [createVNode(Transition, getTransitionProps(mergedMotion.value?.name, mergedMotion.value), { default: () => [withDirectives(createVNode(InternalSubMenuList, { "id": props.id }, { default: () => [slots?.default?.()] }), [[vShow, mergedOpen.value]])] })] });
	};
}, {
	props: {
		id: {
			type: String,
			required: false,
			default: void 0
		},
		open: {
			type: Boolean,
			required: true
		},
		keyPath: {
			type: Array,
			required: true
		}
	},
	name: "InlineSubMenuList",
	inheritAttrs: false
});
//#endregion
export { InlineSubMenuList as default };
