// vxe-pc-ui@4.14.0/lib/components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  install: true,
  loading: true,
  modal: true,
  drawer: true,
  watermark: true,
  print: true,
  saveFile: true,
  readFile: true
};
exports.drawer = void 0;
exports.install = install;
exports.watermark = exports.saveFile = exports.readFile = exports.print = exports.modal = exports.loading = void 0;
var _ui = require("./ui");
Object.keys(_ui).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ui[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ui[key];
    }
  });
});
var _zhCN = _interopRequireDefault(require("./language/zh-CN"));
var _alert = _interopRequireWildcard(require("./alert"));
Object.keys(_alert).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _alert[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _alert[key];
    }
  });
});
var _anchor = _interopRequireWildcard(require("./anchor"));
Object.keys(_anchor).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _anchor[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _anchor[key];
    }
  });
});
var _anchorLink = _interopRequireWildcard(require("./anchor-link"));
Object.keys(_anchorLink).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _anchorLink[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _anchorLink[key];
    }
  });
});
var _avatar = _interopRequireWildcard(require("./avatar"));
Object.keys(_avatar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _avatar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _avatar[key];
    }
  });
});
var _backtop = _interopRequireWildcard(require("./backtop"));
Object.keys(_backtop).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _backtop[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _backtop[key];
    }
  });
});
var _badge = _interopRequireWildcard(require("./badge"));
Object.keys(_badge).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _badge[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _badge[key];
    }
  });
});
var _breadcrumb = _interopRequireWildcard(require("./breadcrumb"));
Object.keys(_breadcrumb).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _breadcrumb[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _breadcrumb[key];
    }
  });
});
var _breadcrumbItem = _interopRequireWildcard(require("./breadcrumb-item"));
Object.keys(_breadcrumbItem).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _breadcrumbItem[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _breadcrumbItem[key];
    }
  });
});
var _button = _interopRequireWildcard(require("./button"));
Object.keys(_button).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _button[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _button[key];
    }
  });
});
var _buttonGroup = _interopRequireWildcard(require("./button-group"));
Object.keys(_buttonGroup).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _buttonGroup[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _buttonGroup[key];
    }
  });
});
var _calendar = _interopRequireWildcard(require("./calendar"));
Object.keys(_calendar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _calendar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _calendar[key];
    }
  });
});
var _card = _interopRequireWildcard(require("./card"));
Object.keys(_card).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _card[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _card[key];
    }
  });
});
var _carousel = _interopRequireWildcard(require("./carousel"));
Object.keys(_carousel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _carousel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _carousel[key];
    }
  });
});
var _carouselItem = _interopRequireWildcard(require("./carousel-item"));
Object.keys(_carouselItem).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _carouselItem[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _carouselItem[key];
    }
  });
});
var _cascader = _interopRequireWildcard(require("./cascader"));
Object.keys(_cascader).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cascader[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _cascader[key];
    }
  });
});
var _checkbox = _interopRequireWildcard(require("./checkbox"));
Object.keys(_checkbox).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _checkbox[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _checkbox[key];
    }
  });
});
var _checkboxButton = _interopRequireWildcard(require("./checkbox-button"));
Object.keys(_checkboxButton).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _checkboxButton[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _checkboxButton[key];
    }
  });
});
var _checkboxGroup = _interopRequireWildcard(require("./checkbox-group"));
Object.keys(_checkboxGroup).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _checkboxGroup[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _checkboxGroup[key];
    }
  });
});
var _col = _interopRequireWildcard(require("./col"));
Object.keys(_col).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _col[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _col[key];
    }
  });
});
var _collapse = _interopRequireWildcard(require("./collapse"));
Object.keys(_collapse).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _collapse[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _collapse[key];
    }
  });
});
var _collapsePane = _interopRequireWildcard(require("./collapse-pane"));
Object.keys(_collapsePane).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _collapsePane[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _collapsePane[key];
    }
  });
});
var _colorPicker = _interopRequireWildcard(require("./color-picker"));
Object.keys(_colorPicker).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _colorPicker[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _colorPicker[key];
    }
  });
});
var _contextMenu = _interopRequireWildcard(require("./context-menu"));
Object.keys(_contextMenu).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _contextMenu[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _contextMenu[key];
    }
  });
});
var _countdown = _interopRequireWildcard(require("./countdown"));
Object.keys(_countdown).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _countdown[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _countdown[key];
    }
  });
});
var _datePanel = _interopRequireWildcard(require("./date-panel"));
Object.keys(_datePanel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datePanel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _datePanel[key];
    }
  });
});
var _datePicker = _interopRequireWildcard(require("./date-picker"));
Object.keys(_datePicker).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datePicker[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _datePicker[key];
    }
  });
});
var _dateRangePicker = _interopRequireWildcard(require("./date-range-picker"));
Object.keys(_dateRangePicker).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dateRangePicker[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _dateRangePicker[key];
    }
  });
});
var _drawer = _interopRequireWildcard(require("./drawer"));
Object.keys(_drawer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _drawer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _drawer[key];
    }
  });
});
var _empty = _interopRequireWildcard(require("./empty"));
Object.keys(_empty).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _empty[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _empty[key];
    }
  });
});
var _form = _interopRequireWildcard(require("./form"));
Object.keys(_form).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _form[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _form[key];
    }
  });
});
var _formGather = _interopRequireWildcard(require("./form-gather"));
Object.keys(_formGather).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formGather[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _formGather[key];
    }
  });
});
var _formGroup = _interopRequireWildcard(require("./form-group"));
Object.keys(_formGroup).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formGroup[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _formGroup[key];
    }
  });
});
var _formItem = _interopRequireWildcard(require("./form-item"));
Object.keys(_formItem).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formItem[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _formItem[key];
    }
  });
});
var _icon = _interopRequireWildcard(require("./icon"));
Object.keys(_icon).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _icon[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _icon[key];
    }
  });
});
var _iconPicker = _interopRequireWildcard(require("./icon-picker"));
Object.keys(_iconPicker).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _iconPicker[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _iconPicker[key];
    }
  });
});
var _image = _interopRequireWildcard(require("./image"));
Object.keys(_image).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _image[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _image[key];
    }
  });
});
var _imageGroup = _interopRequireWildcard(require("./image-group"));
Object.keys(_imageGroup).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _imageGroup[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _imageGroup[key];
    }
  });
});
var _imagePreview = _interopRequireWildcard(require("./image-preview"));
Object.keys(_imagePreview).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _imagePreview[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _imagePreview[key];
    }
  });
});
var _input = _interopRequireWildcard(require("./input"));
Object.keys(_input).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _input[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _input[key];
    }
  });
});
var _layoutAside = _interopRequireWildcard(require("./layout-aside"));
Object.keys(_layoutAside).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _layoutAside[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _layoutAside[key];
    }
  });
});
var _layoutBody = _interopRequireWildcard(require("./layout-body"));
Object.keys(_layoutBody).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _layoutBody[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _layoutBody[key];
    }
  });
});
var _layoutContainer = _interopRequireWildcard(require("./layout-container"));
Object.keys(_layoutContainer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _layoutContainer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _layoutContainer[key];
    }
  });
});
var _layoutFooter = _interopRequireWildcard(require("./layout-footer"));
Object.keys(_layoutFooter).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _layoutFooter[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _layoutFooter[key];
    }
  });
});
var _layoutHeader = _interopRequireWildcard(require("./layout-header"));
Object.keys(_layoutHeader).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _layoutHeader[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _layoutHeader[key];
    }
  });
});
var _link = _interopRequireWildcard(require("./link"));
Object.keys(_link).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _link[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _link[key];
    }
  });
});
var _list = _interopRequireWildcard(require("./list"));
Object.keys(_list).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _list[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _list[key];
    }
  });
});
var _loading = _interopRequireWildcard(require("./loading"));
Object.keys(_loading).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _loading[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _loading[key];
    }
  });
});
var _menu = _interopRequireWildcard(require("./menu"));
Object.keys(_menu).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menu[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _menu[key];
    }
  });
});
var _modal = _interopRequireWildcard(require("./modal"));
Object.keys(_modal).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _modal[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _modal[key];
    }
  });
});
var _noticeBar = _interopRequireWildcard(require("./notice-bar"));
Object.keys(_noticeBar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _noticeBar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _noticeBar[key];
    }
  });
});
var _numberInput = _interopRequireWildcard(require("./number-input"));
Object.keys(_numberInput).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _numberInput[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _numberInput[key];
    }
  });
});
var _optgroup = _interopRequireWildcard(require("./optgroup"));
Object.keys(_optgroup).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _optgroup[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _optgroup[key];
    }
  });
});
var _option = _interopRequireWildcard(require("./option"));
Object.keys(_option).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _option[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _option[key];
    }
  });
});
var _pager = _interopRequireWildcard(require("./pager"));
Object.keys(_pager).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pager[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _pager[key];
    }
  });
});
var _passwordInput = _interopRequireWildcard(require("./password-input"));
Object.keys(_passwordInput).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _passwordInput[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _passwordInput[key];
    }
  });
});
var _printPageBreak = _interopRequireWildcard(require("./print-page-break"));
Object.keys(_printPageBreak).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _printPageBreak[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _printPageBreak[key];
    }
  });
});
var _print = _interopRequireWildcard(require("./print"));
Object.keys(_print).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _print[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _print[key];
    }
  });
});
var _pulldown = _interopRequireWildcard(require("./pulldown"));
Object.keys(_pulldown).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pulldown[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _pulldown[key];
    }
  });
});
var _radio = _interopRequireWildcard(require("./radio"));
Object.keys(_radio).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _radio[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _radio[key];
    }
  });
});
var _radioButton = _interopRequireWildcard(require("./radio-button"));
Object.keys(_radioButton).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _radioButton[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _radioButton[key];
    }
  });
});
var _radioGroup = _interopRequireWildcard(require("./radio-group"));
Object.keys(_radioGroup).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _radioGroup[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _radioGroup[key];
    }
  });
});
var _rate = _interopRequireWildcard(require("./rate"));
Object.keys(_rate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rate[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rate[key];
    }
  });
});
var _result = _interopRequireWildcard(require("./result"));
Object.keys(_result).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _result[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _result[key];
    }
  });
});
var _row = _interopRequireWildcard(require("./row"));
Object.keys(_row).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _row[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _row[key];
    }
  });
});
var _segmented = _interopRequireWildcard(require("./segmented"));
Object.keys(_segmented).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _segmented[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _segmented[key];
    }
  });
});
var _select = _interopRequireWildcard(require("./select"));
Object.keys(_select).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _select[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _select[key];
    }
  });
});
var _splitter = _interopRequireWildcard(require("./splitter"));
Object.keys(_splitter).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _splitter[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _splitter[key];
    }
  });
});
var _splitterPanel = _interopRequireWildcard(require("./splitter-panel"));
Object.keys(_splitterPanel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _splitterPanel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _splitterPanel[key];
    }
  });
});
var _split = _interopRequireWildcard(require("./split"));
Object.keys(_split).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _split[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _split[key];
    }
  });
});
var _splitPane = _interopRequireWildcard(require("./split-pane"));
Object.keys(_splitPane).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _splitPane[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _splitPane[key];
    }
  });
});
var _slider = _interopRequireWildcard(require("./slider"));
Object.keys(_slider).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _slider[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _slider[key];
    }
  });
});
var _steps = _interopRequireWildcard(require("./steps"));
Object.keys(_steps).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _steps[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _steps[key];
    }
  });
});
var _switch = _interopRequireWildcard(require("./switch"));
Object.keys(_switch).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _switch[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _switch[key];
    }
  });
});
var _tabPane = _interopRequireWildcard(require("./tab-pane"));
Object.keys(_tabPane).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tabPane[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tabPane[key];
    }
  });
});
var _tableSelect = _interopRequireWildcard(require("./table-select"));
Object.keys(_tableSelect).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tableSelect[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tableSelect[key];
    }
  });
});
var _tabs = _interopRequireWildcard(require("./tabs"));
Object.keys(_tabs).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tabs[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tabs[key];
    }
  });
});
var _tag = _interopRequireWildcard(require("./tag"));
Object.keys(_tag).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tag[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tag[key];
    }
  });
});
var _textEllipsis = _interopRequireWildcard(require("./text-ellipsis"));
Object.keys(_textEllipsis).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _textEllipsis[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _textEllipsis[key];
    }
  });
});
var _text = _interopRequireWildcard(require("./text"));
Object.keys(_text).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _text[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _text[key];
    }
  });
});
var _textarea = _interopRequireWildcard(require("./textarea"));
Object.keys(_textarea).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _textarea[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _textarea[key];
    }
  });
});
var _timeline = _interopRequireWildcard(require("./timeline"));
Object.keys(_timeline).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _timeline[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _timeline[key];
    }
  });
});
var _timelineItem = _interopRequireWildcard(require("./timeline-item"));
Object.keys(_timelineItem).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _timelineItem[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _timelineItem[key];
    }
  });
});
var _tip = _interopRequireWildcard(require("./tip"));
Object.keys(_tip).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tip[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tip[key];
    }
  });
});
var _tooltip = _interopRequireWildcard(require("./tooltip"));
Object.keys(_tooltip).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tooltip[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tooltip[key];
    }
  });
});
var _tour = _interopRequireWildcard(require("./tour"));
Object.keys(_tour).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tour[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tour[key];
    }
  });
});
var _tree = _interopRequireWildcard(require("./tree"));
Object.keys(_tree).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tree[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tree[key];
    }
  });
});
var _treeSelect = _interopRequireWildcard(require("./tree-select"));
Object.keys(_treeSelect).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _treeSelect[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _treeSelect[key];
    }
  });
});
var _upload = _interopRequireWildcard(require("./upload"));
Object.keys(_upload).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _upload[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _upload[key];
    }
  });
});
var _watermark = _interopRequireWildcard(require("./watermark"));
Object.keys(_watermark).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _watermark[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _watermark[key];
    }
  });
});
var _util = require("./upload/src/util");
var _util2 = require("./print/src/util");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const {
  setI18n,
  setLanguage,
  setTheme,
  setConfig
} = _ui.VxeUI;
const components = [_alert.default, _anchor.default, _anchorLink.default, _avatar.default, _backtop.default, _badge.default, _breadcrumb.default, _breadcrumbItem.default, _button.default, _buttonGroup.default, _calendar.default, _card.default, _carousel.default, _carouselItem.default, _cascader.default, _checkbox.default, _checkboxButton.default, _checkboxGroup.default, _col.default, _collapse.default, _collapsePane.default, _colorPicker.default, _contextMenu.default, _countdown.default, _datePanel.default, _datePicker.default, _dateRangePicker.default, _drawer.default, _empty.default, _form.default, _formGather.default, _formGroup.default, _formItem.default, _icon.default, _iconPicker.default, _image.default, _imageGroup.default, _imagePreview.default, _input.default, _layoutAside.default, _layoutBody.default, _layoutContainer.default, _layoutFooter.default, _layoutHeader.default, _link.default, _list.default, _loading.default, _menu.default, _modal.default, _noticeBar.default, _numberInput.default, _optgroup.default, _option.default, _pager.default, _passwordInput.default, _printPageBreak.default, _print.default, _pulldown.default, _radio.default, _radioButton.default, _radioGroup.default, _rate.default, _result.default, _row.default, _segmented.default, _select.default, _splitter.default, _splitterPanel.default, _split.default, _splitPane.default, _slider.default, _steps.default, _switch.default, _tabPane.default, _tableSelect.default, _tabs.default, _tag.default, _textEllipsis.default, _text.default, _textarea.default, _timeline.default, _timelineItem.default, _tip.default, _tooltip.default, _tour.default, _tree.default, _treeSelect.default, _upload.default, _watermark.default];
function install(app, options) {
  setConfig(options);
  components.forEach(component => app.use(component));
}
// 默认中文
const defaultLanguage = 'zh-CN';
setI18n(defaultLanguage, _zhCN.default);
setLanguage(defaultLanguage);
setTheme('light');
// 兼容老版本
const loading = exports.loading = _loading.LoadingController;
const modal = exports.modal = _modal.ModalController;
const drawer = exports.drawer = _drawer.DrawerController;
const watermark = exports.watermark = _watermark.WatermarkController;
const print = exports.print = _util2.printHtml;
const saveFile = exports.saveFile = _util.saveLocalFile;
const readFile = exports.readFile = _util.readLocalFile;

// Components
