// @v-c/qrcode@1.0.0/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_utils = require("./utils.cjs");
const require_QRCodeCanvas = require("./QRCodeCanvas.cjs");
const require_QRCodeSVG = require("./QRCodeSVG.cjs");
exports.DEFAULT_BACKGROUND_COLOR = require_utils.DEFAULT_BACKGROUND_COLOR;
exports.DEFAULT_FRONT_COLOR = require_utils.DEFAULT_FRONT_COLOR;
exports.DEFAULT_IMG_SCALE = require_utils.DEFAULT_IMG_SCALE;
exports.DEFAULT_LEVEL = require_utils.DEFAULT_LEVEL;
exports.DEFAULT_MARGIN_SIZE = require_utils.DEFAULT_MARGIN_SIZE;
exports.DEFAULT_MINVERSION = require_utils.DEFAULT_MINVERSION;
exports.DEFAULT_NEED_MARGIN = require_utils.DEFAULT_NEED_MARGIN;
exports.DEFAULT_SIZE = require_utils.DEFAULT_SIZE;
exports.ERROR_LEVEL_MAP = require_utils.ERROR_LEVEL_MAP;
exports.QRCodeCanvas = require_QRCodeCanvas.QRCodeCanvas;
exports.QRCodeSVG = require_QRCodeSVG.QRCodeSVG;
exports.SPEC_MARGIN_SIZE = require_utils.SPEC_MARGIN_SIZE;
exports.excavateModules = require_utils.excavateModules;
exports.generatePath = require_utils.generatePath;
exports.getImageSettings = require_utils.getImageSettings;
exports.getMarginSize = require_utils.getMarginSize;
exports.isSupportPath2d = require_utils.isSupportPath2d;
