// @v-c/menu@1.0.14/dist/context/PathContext.js
import { computed, defineComponent, inject, provide, ref } from "vue";
//#region src/context/PathContext.tsx
var EmptyList = [];
var PathRegisterContextKey = Symbol("PathRegisterContext");
function useMeasure() {
	return inject(PathRegisterContextKey, null);
}
function useMeasureProvider(context) {
	provide(PathRegisterContextKey, context);
}
var MeasureProvider = /* @__PURE__ */ defineComponent((props, { slots }) => {
	useMeasureProvider(props);
	return () => {
		return slots?.default?.();
	};
}, { props: {
	registerPath: {
		type: Function,
		required: true
	},
	unregisterPath: {
		type: Function,
		required: true
	}
} });
var PathTrackerContextKey = Symbol("PathTrackerContext");
function useFullPath(eventKey) {
	const parentKeyPath = inject(PathTrackerContextKey, ref(EmptyList));
	return computed(() => {
		return eventKey !== void 0 ? [...parentKeyPath.value, eventKey.value] : parentKeyPath.value;
	});
}
var PathTrackerProvider = /* @__PURE__ */ defineComponent((props, { slots }) => {
	provide(PathTrackerContextKey, computed(() => props.value));
	return () => {
		return slots?.default?.();
	};
}, { props: { value: {
	type: Array,
	required: true
} } });
var PathUserContextKey = Symbol("PathUserContext");
function usePathUserContextProvider(context) {
	provide(PathUserContextKey, context);
}
function usePathUserContext() {
	return inject(PathUserContextKey, { isSubPathKey: () => false });
}
var PathUserProvider = /* @__PURE__ */ defineComponent((props, { slots }) => {
	usePathUserContextProvider(props);
	return () => {
		return slots?.default?.();
	};
}, { props: { isSubPathKey: {
	type: Function,
	required: true
} } });
var PathTrackerContext = { Provider: PathTrackerProvider };
var PathRegisterContext = { Provider: MeasureProvider };
var PathUserContext = { Provider: PathUserProvider };
//#endregion
export { MeasureProvider, PathRegisterContext, PathTrackerContext, PathTrackerProvider, PathUserContext, PathUserProvider, useFullPath, useMeasure, useMeasureProvider, usePathUserContext, usePathUserContextProvider };
