// antdv-next@1.2.2/dist/statistic/Number.js
import { createVNode, defineComponent } from "vue";

//#region src/statistic/Number.tsx
const StatisticNumber = /* @__PURE__ */ defineComponent((props) => {
	return () => {
		const { value, formatter, precision, decimalSeparator, groupSeparator = "", prefixCls, className, style } = props;
		let valueNode;
		if (typeof formatter === "function") valueNode = formatter(value);
		else {
			const val = String(value);
			const cells = val.match(/^(-?)(\d*)(\.(\d+))?$/);
			if (!cells || val === "-") valueNode = val;
			else {
				const negative = cells[1];
				let int = cells[2] || "0";
				let decimal = cells[4] || "";
				int = int.replace(/\B(?=(\d{3})+(?!\d))/g, groupSeparator);
				if (typeof precision === "number") decimal = decimal.padEnd(precision, "0").slice(0, precision > 0 ? precision : 0);
				if (decimal) decimal = `${decimalSeparator}${decimal}`;
				valueNode = [createVNode("span", {
					"key": "int",
					"class": `${prefixCls}-content-value-int`
				}, [negative, int]), decimal && createVNode("span", {
					"key": "decimal",
					"class": `${prefixCls}-content-value-decimal`
				}, [decimal])];
			}
		}
		return createVNode("span", {
			"class": className,
			"style": style
		}, [valueNode]);
	};
}, { props: {
	value: {
		type: [Number, String],
		required: true
	},
	prefixCls: {
		type: String,
		required: false
	},
	className: {
		type: String,
		required: false
	},
	style: {
		type: Object,
		required: false
	},
	formatter: {
		type: [
			Boolean,
			String,
			Function
		],
		required: false,
		default: void 0
	},
	decimalSeparator: {
		type: String,
		required: false
	},
	groupSeparator: {
		type: String,
		required: false
	},
	precision: {
		type: Number,
		required: false
	}
} });
var Number_default = StatisticNumber;

//#endregion
export { Number_default as default };
