// @v-c/table@1.0.7/dist/sugar/Column.js
import { defineComponent } from "vue";
//#region src/sugar/Column.tsx
/**
* This is a syntactic sugar for `columns` prop.
* So HOC will not work on this.
*/
var Column = /* @__PURE__ */ defineComponent(() => {
	return () => null;
}, { props: {
	colSpan: {
		type: Number,
		required: false,
		default: void 0
	},
	dataIndex: {
		required: false,
		default: void 0
	},
	render: {
		type: Function,
		required: false,
		default: void 0
	},
	shouldCellUpdate: {
		type: Function,
		required: false,
		default: void 0
	},
	rowSpan: {
		type: Number,
		required: false,
		default: void 0
	},
	width: {
		type: [Number, String],
		required: false,
		default: void 0
	},
	minWidth: {
		type: Number,
		required: false,
		default: void 0
	},
	onCell: {
		type: Function,
		required: false,
		default: void 0
	},
	onCellClick: {
		type: Function,
		required: false,
		default: void 0
	},
	title: {
		type: [
			Object,
			Function,
			String,
			Number,
			null,
			Boolean,
			Array
		],
		required: false,
		default: void 0
	},
	key: {
		type: [String, Number],
		required: false,
		default: void 0
	},
	className: {
		type: String,
		required: false,
		default: void 0
	},
	hidden: {
		type: Boolean,
		required: false,
		default: void 0
	},
	fixed: {
		type: [String, Boolean],
		required: false,
		default: void 0
	},
	onHeaderCell: {
		type: Function,
		required: false,
		default: void 0
	},
	ellipsis: {
		type: [Object, Boolean],
		required: false,
		default: void 0
	},
	align: {
		type: String,
		required: false,
		default: void 0
	},
	rowScope: {
		type: String,
		required: false,
		default: void 0
	}
} });
//#endregion
export { Column as default };
