// reka-ui@2.9.6/dist/shared.cjs
const require_shared_arrays = require('./shared/arrays.cjs');
const require_shared_browser = require('./shared/browser.cjs');
const require_shared_clamp = require('./shared/clamp.cjs');
const require_shared_createContext = require('./shared/createContext.cjs');
const require_shared_getActiveElement = require('./shared/getActiveElement.cjs');
const require_shared_handleAndDispatchCustomEvent = require('./shared/handleAndDispatchCustomEvent.cjs');
const require_shared_isValidVNodeElement = require('./shared/isValidVNodeElement.cjs');
const require_shared_nullish = require('./shared/nullish.cjs');
const require_shared_isValueEqualOrExist = require('./shared/isValueEqualOrExist.cjs');
const require_shared_object = require('./shared/object.cjs');
const require_shared_onFocusOutside = require('./shared/onFocusOutside.cjs');
const require_shared_renderSlotFragments = require('./shared/renderSlotFragments.cjs');
const require_shared_trap_focus = require('./shared/trap-focus.cjs');
const require_shared_useArrowNavigation = require('./shared/useArrowNavigation.cjs');
require('./ConfigProvider/ConfigProvider.cjs');
const require_shared_useBodyScrollLock = require('./shared/useBodyScrollLock.cjs');
require('./date/comparators.cjs');
require('./date/utils.cjs');
require('./date/calendar.cjs');
require('./date/index.cjs');
const require_shared_useDateFormatter = require('./shared/useDateFormatter.cjs');
const require_shared_useDirection = require('./shared/useDirection.cjs');
const require_shared_useEmitAsProps = require('./shared/useEmitAsProps.cjs');
const require_shared_useFilter = require('./shared/useFilter.cjs');
const require_shared_useFocusGuards = require('./shared/useFocusGuards.cjs');
const require_shared_useFormControl = require('./shared/useFormControl.cjs');
const require_shared_useForwardExpose = require('./shared/useForwardExpose.cjs');
const require_shared_useForwardProps = require('./shared/useForwardProps.cjs');
const require_shared_useForwardPropsEmits = require('./shared/useForwardPropsEmits.cjs');
const require_shared_useForwardRef = require('./shared/useForwardRef.cjs');
const require_shared_useGraceArea = require('./shared/useGraceArea.cjs');
const require_shared_useHideOthers = require('./shared/useHideOthers.cjs');
const require_shared_useId = require('./shared/useId.cjs');
const require_shared_useKbd = require('./shared/useKbd.cjs');
const require_shared_useLocale = require('./shared/useLocale.cjs');
const require_shared_useSelectionBehavior = require('./shared/useSelectionBehavior.cjs');
const require_shared_useSize = require('./shared/useSize.cjs');
const require_shared_useStateMachine = require('./shared/useStateMachine.cjs');
const require_shared_useTypeahead = require('./shared/useTypeahead.cjs');
const require_shared_withDefault = require('./shared/withDefault.cjs');
require('./shared/index.cjs');

exports.areEqual = require_shared_arrays.areEqual;
exports.chunk = require_shared_arrays.chunk;
exports.clamp = require_shared_clamp.clamp;
exports.createContext = require_shared_createContext.createContext;
exports.findValuesBetween = require_shared_arrays.findValuesBetween;
exports.getActiveElement = require_shared_getActiveElement.getActiveElement;
exports.handleAndDispatchCustomEvent = require_shared_handleAndDispatchCustomEvent.handleAndDispatchCustomEvent;
exports.isBrowser = require_shared_browser.isBrowser;
exports.isNullish = require_shared_nullish.isNullish;
exports.isValidVNodeElement = require_shared_isValidVNodeElement.isValidVNodeElement;
exports.isValueEqualOrExist = require_shared_isValueEqualOrExist.isValueEqualOrExist;
exports.omit = require_shared_object.omit;
exports.onFocusOutside = require_shared_onFocusOutside.onFocusOutside;
exports.pick = require_shared_object.pick;
exports.renderSlotFragments = require_shared_renderSlotFragments.renderSlotFragments;
exports.roundToStepPrecision = require_shared_clamp.roundToStepPrecision;
exports.snapValueToStep = require_shared_clamp.snapValueToStep;
exports.trapFocus = require_shared_trap_focus.trapFocus;
exports.useArrowNavigation = require_shared_useArrowNavigation.useArrowNavigation;
exports.useBodyScrollLock = require_shared_useBodyScrollLock.useBodyScrollLock;
exports.useDateFormatter = require_shared_useDateFormatter.useDateFormatter;
exports.useDirection = require_shared_useDirection.useDirection;
exports.useEmitAsProps = require_shared_useEmitAsProps.useEmitAsProps;
exports.useFilter = require_shared_useFilter.useFilter;
exports.useFocusGuards = require_shared_useFocusGuards.useFocusGuards;
exports.useFormControl = require_shared_useFormControl.useFormControl;
exports.useForwardExpose = require_shared_useForwardExpose.useForwardExpose;
exports.useForwardProps = require_shared_useForwardProps.useForwardProps;
exports.useForwardPropsEmits = require_shared_useForwardPropsEmits.useForwardPropsEmits;
exports.useForwardRef = require_shared_useForwardRef.useForwardRef;
exports.useGraceArea = require_shared_useGraceArea.useGraceArea;
exports.useHideOthers = require_shared_useHideOthers.useHideOthers;
exports.useId = require_shared_useId.useId;
exports.useKbd = require_shared_useKbd.useKbd;
exports.useLocale = require_shared_useLocale.useLocale;
exports.useSelectionBehavior = require_shared_useSelectionBehavior.useSelectionBehavior;
exports.useSize = require_shared_useSize.useSize;
exports.useStateMachine = require_shared_useStateMachine.useStateMachine;
exports.useTestKbd = require_shared_useKbd.useTestKbd;
exports.useTypeahead = require_shared_useTypeahead.useTypeahead;
exports.withDefault = require_shared_withDefault.withDefault;
