// reka-ui@2.9.6/dist/shared.js
import { areEqual, chunk, findValuesBetween } from "./shared/arrays.js";
import { isBrowser } from "./shared/browser.js";
import { clamp, roundToStepPrecision, snapValueToStep } from "./shared/clamp.js";
import { createContext } from "./shared/createContext.js";
import { getActiveElement } from "./shared/getActiveElement.js";
import { handleAndDispatchCustomEvent } from "./shared/handleAndDispatchCustomEvent.js";
import { isValidVNodeElement } from "./shared/isValidVNodeElement.js";
import { isNullish } from "./shared/nullish.js";
import { isValueEqualOrExist } from "./shared/isValueEqualOrExist.js";
import { omit, pick } from "./shared/object.js";
import { onFocusOutside } from "./shared/onFocusOutside.js";
import { renderSlotFragments } from "./shared/renderSlotFragments.js";
import { trapFocus } from "./shared/trap-focus.js";
import { useArrowNavigation } from "./shared/useArrowNavigation.js";
import "./ConfigProvider/ConfigProvider.js";
import { useBodyScrollLock } from "./shared/useBodyScrollLock.js";
import "./date/comparators.js";
import "./date/utils.js";
import "./date/calendar.js";
import "./date/index.js";
import { useDateFormatter } from "./shared/useDateFormatter.js";
import { useDirection } from "./shared/useDirection.js";
import { useEmitAsProps } from "./shared/useEmitAsProps.js";
import { useFilter } from "./shared/useFilter.js";
import { useFocusGuards } from "./shared/useFocusGuards.js";
import { useFormControl } from "./shared/useFormControl.js";
import { useForwardExpose } from "./shared/useForwardExpose.js";
import { useForwardProps } from "./shared/useForwardProps.js";
import { useForwardPropsEmits } from "./shared/useForwardPropsEmits.js";
import { useForwardRef } from "./shared/useForwardRef.js";
import { useGraceArea } from "./shared/useGraceArea.js";
import { useHideOthers } from "./shared/useHideOthers.js";
import { useId } from "./shared/useId.js";
import { useKbd, useTestKbd } from "./shared/useKbd.js";
import { useLocale } from "./shared/useLocale.js";
import { useSelectionBehavior } from "./shared/useSelectionBehavior.js";
import { useSize } from "./shared/useSize.js";
import { useStateMachine } from "./shared/useStateMachine.js";
import { useTypeahead } from "./shared/useTypeahead.js";
import { withDefault } from "./shared/withDefault.js";
import "./shared/index.js";

export { areEqual, chunk, clamp, createContext, findValuesBetween, getActiveElement, handleAndDispatchCustomEvent, isBrowser, isNullish, isValidVNodeElement, isValueEqualOrExist, omit, onFocusOutside, pick, renderSlotFragments, roundToStepPrecision, snapValueToStep, trapFocus, useArrowNavigation, useBodyScrollLock, useDateFormatter, useDirection, useEmitAsProps, useFilter, useFocusGuards, useFormControl, useForwardExpose, useForwardProps, useForwardPropsEmits, useForwardRef, useGraceArea, useHideOthers, useId, useKbd, useLocale, useSelectionBehavior, useSize, useStateMachine, useTestKbd, useTypeahead, withDefault };
