// antdv-next@1.2.2/dist/collapse/CollapsePanel.js
import { defineComponent } from "vue";

//#region src/collapse/CollapsePanel.tsx
const COLLAPSE_PANEL_MARK = "_ANTDV_NEXT_COLLAPSE_PANEL";
const CollapsePanel = /* @__PURE__ */ defineComponent(() => {
	return () => null;
}, {
	props: {
		key: {
			type: [String, Number],
			required: true
		},
		class: {
			type: String,
			required: false
		},
		style: {
			type: Object,
			required: false
		},
		header: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		showArrow: {
			type: Boolean,
			required: false,
			default: void 0
		},
		prefixCls: {
			type: String,
			required: false
		},
		forceRender: {
			type: Boolean,
			required: false,
			default: void 0
		},
		id: {
			type: String,
			required: false
		},
		extra: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		collapsible: {
			type: String,
			required: false
		},
		classes: {
			type: Object,
			required: false
		},
		styles: {
			type: Object,
			required: false
		}
	},
	name: "ACollapsePanel",
	inheritAttrs: false
});
CollapsePanel[COLLAPSE_PANEL_MARK] = true;
var CollapsePanel_default = CollapsePanel;

//#endregion
export { COLLAPSE_PANEL_MARK, CollapsePanel_default as default };
