// antdv-next@1.2.2/dist/theme/themes/dark/colors.js
import { getAlphaColor, getSolidColor } from "./colorAlgorithm.js";
import { generate } from "@ant-design/colors";

//#region src/theme/themes/dark/colors.ts
const generateColorPalettes = (baseColor) => {
	const colors = generate(baseColor, { theme: "dark" });
	return {
		1: colors[0],
		2: colors[1],
		3: colors[2],
		4: colors[3],
		5: colors[6],
		6: colors[5],
		7: colors[4],
		8: colors[6],
		9: colors[5],
		10: colors[4]
	};
};
const generateNeutralColorPalettes = (bgBaseColor, textBaseColor, shadowColor) => {
	const colorBgBase = bgBaseColor || "#000";
	const colorTextBase = textBaseColor || "#fff";
	return {
		colorBgBase,
		colorTextBase,
		colorShadow: shadowColor || "rgba(255, 255, 255, 0.2)",
		colorText: getAlphaColor(colorTextBase, .85),
		colorTextSecondary: getAlphaColor(colorTextBase, .65),
		colorTextTertiary: getAlphaColor(colorTextBase, .45),
		colorTextQuaternary: getAlphaColor(colorTextBase, .25),
		colorFill: getAlphaColor(colorTextBase, .18),
		colorFillSecondary: getAlphaColor(colorTextBase, .12),
		colorFillTertiary: getAlphaColor(colorTextBase, .08),
		colorFillQuaternary: getAlphaColor(colorTextBase, .04),
		colorBgSolid: getAlphaColor(colorTextBase, .95),
		colorBgSolidHover: getAlphaColor(colorTextBase, 1),
		colorBgSolidActive: getAlphaColor(colorTextBase, .9),
		colorBgElevated: getSolidColor(colorBgBase, 12),
		colorBgContainer: getSolidColor(colorBgBase, 8),
		colorBgLayout: getSolidColor(colorBgBase, 0),
		colorBgSpotlight: getSolidColor(colorBgBase, 26),
		colorBgBlur: getAlphaColor(colorTextBase, .04),
		colorBorder: getSolidColor(colorBgBase, 26),
		colorBorderDisabled: getSolidColor(colorBgBase, 26),
		colorBorderSecondary: getSolidColor(colorBgBase, 19)
	};
};

//#endregion
export { generateColorPalettes, generateNeutralColorPalettes };
