// antdv-next@1.2.2/dist/descriptions/Item.js
import { defineComponent } from "vue";

//#region src/descriptions/Item.tsx
const DESCRIPTIONS_ITEM_MARK = "_ANTDV_NEXT_DESCRIPTIONS_ITEM";
const DescriptionsItem = /* @__PURE__ */ defineComponent(() => {
	return () => null;
}, {
	props: {
		class: {
			type: String,
			required: false
		},
		style: {
			type: Object,
			required: false
		},
		label: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		classes: {
			type: Object,
			required: false
		},
		styles: {
			type: Object,
			required: false
		},
		content: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		span: {
			type: [
				Number,
				String,
				Object
			],
			required: false
		},
		rootClass: {
			type: String,
			required: false
		},
		prefixCls: {
			type: String,
			required: false
		}
	},
	name: "ADescriptionsItem",
	inheritAttrs: false
});
DescriptionsItem[DESCRIPTIONS_ITEM_MARK] = true;
var Item_default = DescriptionsItem;

//#endregion
export { DESCRIPTIONS_ITEM_MARK, Item_default as default };
