// antdv-next@1.2.2/dist/input/style/index.js
import { clearFix, genFocusOutline, resetComponent } from "../../style/index.js";
import { genStyleHooks } from "../../theme/util/genStyleUtils.js";
import { mergeToken } from "../../theme/internal.js";
import { genCompactItemStyle } from "../../style/compact-item.js";
import { initComponentToken, initInputToken } from "./token.js";
import { genBorderlessStyle, genFilledGroupStyle, genFilledStyle, genOutlinedGroupStyle, genOutlinedStyle, genUnderlinedStyle } from "./variants.js";
import { unit } from "@antdv-next/cssinjs";

//#region src/input/style/index.ts
function genPlaceholderStyle(color) {
	return {
		"&::-moz-placeholder": { opacity: 1 },
		"&::placeholder": {
			color,
			userSelect: "none"
		},
		"&:placeholder-shown": { textOverflow: "ellipsis" }
	};
}
const genActiveStyle = (token) => ({
	borderColor: token.activeBorderColor,
	boxShadow: token.activeShadow,
	outline: 0,
	backgroundColor: token.activeBg
});
const genInputLargeStyle = (token) => {
	const { paddingBlockLG, lineHeightLG, borderRadiusLG, paddingInlineLG } = token;
	return {
		padding: `${unit(paddingBlockLG)} ${unit(paddingInlineLG)}`,
		fontSize: token.inputFontSizeLG,
		lineHeight: lineHeightLG,
		borderRadius: borderRadiusLG
	};
};
const genInputSmallStyle = (token) => ({
	padding: `${unit(token.paddingBlockSM)} ${unit(token.paddingInlineSM)}`,
	fontSize: token.inputFontSizeSM,
	borderRadius: token.borderRadiusSM
});
function genBasicInputStyle(token, option = {}) {
	return {
		position: "relative",
		display: "inline-block",
		width: "100%",
		minWidth: 0,
		padding: `${unit(token.paddingBlock)} ${unit(token.paddingInline)}`,
		color: token.colorText,
		fontSize: token.inputFontSize,
		lineHeight: token.lineHeight,
		borderRadius: token.borderRadius,
		transition: `all ${token.motionDurationMid}`,
		...genPlaceholderStyle(token.colorTextPlaceholder),
		"&-lg": {
			...genInputLargeStyle(token),
			...option.largeStyle
		},
		"&-sm": {
			...genInputSmallStyle(token),
			...option.smallStyle
		},
		"&-rtl, &-textarea-rtl": { direction: "rtl" }
	};
}
const genInputGroupStyle = (token) => {
	const { componentCls, antCls } = token;
	return {
		position: "relative",
		display: "table",
		width: "100%",
		borderCollapse: "separate",
		borderSpacing: 0,
		"&[class*='col-']": {
			paddingInlineEnd: token.paddingXS,
			"&:last-child": { paddingInlineEnd: 0 }
		},
		[`&-lg ${componentCls}, &-lg > ${componentCls}-group-addon`]: { ...genInputLargeStyle(token) },
		[`&-sm ${componentCls}, &-sm > ${componentCls}-group-addon`]: { ...genInputSmallStyle(token) },
		[`&-lg ${antCls}-select-single`]: { height: token.controlHeightLG },
		[`&-sm ${antCls}-select-single`]: { height: token.controlHeightSM },
		[`> ${componentCls}`]: {
			display: "table-cell",
			"&:not(:first-child):not(:last-child)": { borderRadius: 0 }
		},
		[`${componentCls}-group`]: {
			"&-addon, &-wrap": {
				display: "table-cell",
				width: 1,
				whiteSpace: "nowrap",
				verticalAlign: "middle",
				"&:not(:first-child):not(:last-child)": { borderRadius: 0 }
			},
			"&-wrap > *": { display: "block !important" },
			"&-addon": {
				position: "relative",
				padding: `0 ${unit(token.paddingInline)}`,
				color: token.colorText,
				fontWeight: "normal",
				fontSize: token.inputFontSize,
				textAlign: "center",
				borderRadius: token.borderRadius,
				transition: `all ${token.motionDurationSlow}`,
				lineHeight: 1,
				[`${antCls}-select`]: {
					margin: `${unit(token.calc(token.paddingBlock).add(1).mul(-1).equal())} ${unit(token.calc(token.paddingInline).mul(-1).equal())}`,
					[`&${antCls}-select-single:not(${antCls}-select-customize-input):not(${antCls}-pagination-size-changer)`]: {
						backgroundColor: "inherit",
						border: `${unit(token.lineWidth)} ${token.lineType} transparent`,
						boxShadow: "none"
					}
				},
				[`${antCls}-cascader-picker`]: {
					margin: `-9px ${unit(token.calc(token.paddingInline).mul(-1).equal())}`,
					backgroundColor: "transparent",
					[`${antCls}-cascader-input`]: {
						textAlign: "start",
						border: 0,
						boxShadow: "none"
					}
				}
			}
		},
		[componentCls]: {
			width: "100%",
			marginBottom: 0,
			textAlign: "inherit",
			"&:focus": {
				zIndex: 1,
				borderInlineEndWidth: 1
			},
			"&:hover": {
				zIndex: 1,
				borderInlineEndWidth: 1
			}
		},
		[`> ${componentCls}:first-child, ${componentCls}-group-addon:first-child`]: {
			borderStartEndRadius: 0,
			borderEndEndRadius: 0,
			[`${antCls}-select`]: {
				borderStartEndRadius: 0,
				borderEndEndRadius: 0
			}
		},
		[`> ${componentCls}-affix-wrapper`]: {
			[`&:not(:first-child) ${componentCls}`]: {
				borderStartStartRadius: 0,
				borderEndStartRadius: 0
			},
			[`&:not(:last-child) ${componentCls}`]: {
				borderStartEndRadius: 0,
				borderEndEndRadius: 0
			}
		},
		[`> ${componentCls}:last-child, ${componentCls}-group-addon:last-child`]: {
			borderStartStartRadius: 0,
			borderEndStartRadius: 0,
			[`${antCls}-select`]: {
				borderStartStartRadius: 0,
				borderEndStartRadius: 0
			}
		},
		[`${componentCls}-affix-wrapper`]: {
			"&:not(:last-child)": {
				borderStartEndRadius: 0,
				borderEndEndRadius: 0
			},
			"&:not(:first-child)": {
				borderStartStartRadius: 0,
				borderEndStartRadius: 0
			}
		},
		[`&${componentCls}-group-compact`]: {
			display: "block",
			...clearFix(),
			[`${componentCls}-group-addon, ${componentCls}-group-wrap, > ${componentCls}`]: { "&:not(:first-child):not(:last-child)": {
				borderInlineEndWidth: token.lineWidth,
				"&:hover, &:focus": { zIndex: 1 }
			} },
			"& > *": {
				display: "inline-flex",
				float: "none",
				verticalAlign: "top",
				borderRadius: 0
			},
			[`
        & > ${componentCls}-affix-wrapper,
        & > ${componentCls}-number-affix-wrapper,
        & > ${antCls}-picker-range
      `]: { display: "inline-flex" },
			"& > *:not(:last-child)": {
				marginInlineEnd: token.calc(token.lineWidth).mul(-1).equal(),
				borderInlineEndWidth: token.lineWidth
			},
			[componentCls]: { float: "none" },
			[`& > ${antCls}-select,
      & > ${antCls}-select-auto-complete ${componentCls},
      & > ${antCls}-cascader-picker ${componentCls},
      & > ${componentCls}-group-wrapper ${componentCls}`]: {
				borderInlineEndWidth: token.lineWidth,
				borderRadius: 0,
				"&:hover, &:focus": { zIndex: 1 }
			},
			[`& > ${antCls}-select-focused`]: { zIndex: 1 },
			[`& > ${antCls}-select > ${antCls}-select-arrow`]: { zIndex: 1 },
			[`& > *:first-child,
      & > ${antCls}-select:first-child,
      & > ${antCls}-select-auto-complete:first-child ${componentCls},
      & > ${antCls}-cascader-picker:first-child ${componentCls}`]: {
				borderStartStartRadius: token.borderRadius,
				borderEndStartRadius: token.borderRadius
			},
			[`& > *:last-child,
      & > ${antCls}-select:last-child,
      & > ${antCls}-cascader-picker:last-child ${componentCls},
      & > ${antCls}-cascader-picker-focused:last-child ${componentCls}`]: {
				borderInlineEndWidth: token.lineWidth,
				borderStartEndRadius: token.borderRadius,
				borderEndEndRadius: token.borderRadius
			},
			[`& > ${antCls}-select-auto-complete ${componentCls}`]: { verticalAlign: "top" },
			[`${componentCls}-group-wrapper + ${componentCls}-group-wrapper`]: {
				marginInlineStart: token.calc(token.lineWidth).mul(-1).equal(),
				[`${componentCls}-affix-wrapper`]: {}
			}
		}
	};
};
const genInputStyle = (token) => {
	const { componentCls, controlHeightSM, lineWidth, calc } = token;
	const colorSmallPadding = calc(controlHeightSM).sub(calc(lineWidth).mul(2)).sub(16).div(2).equal();
	return { [componentCls]: {
		...resetComponent(token),
		...genBasicInputStyle(token),
		...genOutlinedStyle(token),
		...genFilledStyle(token),
		...genBorderlessStyle(token),
		...genUnderlinedStyle(token),
		"&[type=\"color\"]": {
			height: token.controlHeight,
			[`&${componentCls}-lg`]: { height: token.controlHeightLG },
			[`&${componentCls}-sm`]: {
				height: controlHeightSM,
				paddingTop: colorSmallPadding,
				paddingBottom: colorSmallPadding
			}
		},
		"&[type=\"search\"]::-webkit-search-cancel-button, &[type=\"search\"]::-webkit-search-decoration": { appearance: "none" }
	} };
};
const genAllowClearStyle = (token) => {
	const { componentCls } = token;
	return { [`${componentCls}-clear-icon`]: {
		margin: 0,
		padding: 0,
		lineHeight: 0,
		color: token.colorTextQuaternary,
		fontSize: token.fontSizeIcon,
		verticalAlign: -1,
		cursor: "pointer",
		transition: `color ${token.motionDurationSlow}`,
		border: "none",
		outline: "none",
		backgroundColor: "transparent",
		"&:hover": { color: token.colorIcon },
		"&:focus-visible": {
			color: token.colorIcon,
			borderRadius: token.borderRadiusSM,
			...genFocusOutline(token)
		},
		"&:active": { color: token.colorText },
		"&-hidden": { visibility: "hidden" },
		"&-has-suffix": { margin: `0 ${unit(token.inputAffixPadding)}` }
	} };
};
const genAffixStyle = (token) => {
	const { componentCls, inputAffixPadding, colorTextDescription, motionDurationSlow, colorIcon, colorIconHover, iconCls } = token;
	const affixCls = `${componentCls}-affix-wrapper`;
	const affixClsDisabled = `${componentCls}-affix-wrapper-disabled`;
	return {
		[affixCls]: {
			...genBasicInputStyle(token),
			display: "inline-flex",
			"&-focused, &:focus": { zIndex: 1 },
			[`> input${componentCls}`]: { padding: 0 },
			[`> input${componentCls}, > textarea${componentCls}`]: {
				fontSize: "inherit",
				border: "none",
				borderRadius: 0,
				outline: "none",
				background: "transparent",
				color: "inherit",
				"&::-ms-reveal": { display: "none" },
				"&:focus": { boxShadow: "none !important" }
			},
			"&::before": {
				display: "inline-block",
				width: 0,
				visibility: "hidden",
				content: "\"\\a0\""
			},
			[componentCls]: {
				"&-prefix, &-suffix": {
					display: "flex",
					flex: "none",
					alignItems: "center",
					"> *:not(:last-child)": { marginInlineEnd: token.paddingXS }
				},
				"&-show-count-suffix": {
					color: colorTextDescription,
					direction: "ltr"
				},
				"&-show-count-has-suffix": { marginInlineEnd: token.paddingXXS },
				"&-prefix": { marginInlineEnd: inputAffixPadding },
				"&-suffix": { marginInlineStart: inputAffixPadding }
			},
			...genAllowClearStyle(token),
			[`${iconCls}${componentCls}-password-icon`]: {
				color: colorIcon,
				cursor: "pointer",
				transition: `all ${motionDurationSlow}`,
				"&:hover": { color: colorIconHover }
			}
		},
		[`${componentCls}-underlined`]: { borderRadius: 0 },
		[affixClsDisabled]: { [`${iconCls}${componentCls}-password-icon`]: {
			color: colorIcon,
			cursor: "not-allowed",
			"&:hover": { color: colorIcon }
		} }
	};
};
const genGroupStyle = (token) => {
	const { componentCls, borderRadiusLG, borderRadiusSM } = token;
	return { [`${componentCls}-group`]: {
		...resetComponent(token),
		...genInputGroupStyle(token),
		"&-rtl": { direction: "rtl" },
		"&-wrapper": {
			display: "inline-block",
			width: "100%",
			textAlign: "start",
			verticalAlign: "top",
			"&-rtl": { direction: "rtl" },
			"&-lg": { [`${componentCls}-group-addon`]: {
				borderRadius: borderRadiusLG,
				fontSize: token.inputFontSizeLG
			} },
			"&-sm": { [`${componentCls}-group-addon`]: { borderRadius: borderRadiusSM } },
			...genOutlinedGroupStyle(token),
			...genFilledGroupStyle(token),
			[`&:not(${componentCls}-compact-first-item):not(${componentCls}-compact-last-item)${componentCls}-compact-item`]: { [`${componentCls}, ${componentCls}-group-addon`]: { borderRadius: 0 } },
			[`&:not(${componentCls}-compact-last-item)${componentCls}-compact-first-item`]: { [`${componentCls}, ${componentCls}-group-addon`]: {
				borderStartEndRadius: 0,
				borderEndEndRadius: 0
			} },
			[`&:not(${componentCls}-compact-first-item)${componentCls}-compact-last-item`]: { [`${componentCls}, ${componentCls}-group-addon`]: {
				borderStartStartRadius: 0,
				borderEndStartRadius: 0
			} },
			[`&:not(${componentCls}-compact-last-item)${componentCls}-compact-item`]: { [`${componentCls}-affix-wrapper`]: {
				borderStartEndRadius: 0,
				borderEndEndRadius: 0
			} },
			[`&:not(${componentCls}-compact-first-item)${componentCls}-compact-item`]: { [`${componentCls}-affix-wrapper`]: {
				borderStartStartRadius: 0,
				borderEndStartRadius: 0
			} }
		}
	} };
};
const genRangeStyle = (token) => {
	const { componentCls } = token;
	return { [`${componentCls}-out-of-range`]: { [`&, & input, & textarea, ${componentCls}-show-count-suffix, ${componentCls}-data-count`]: { color: token.colorError } } };
};
const useSharedStyle = genStyleHooks(["Input", "Shared"], (token) => {
	const inputToken = mergeToken(token, initInputToken(token));
	return [genInputStyle(inputToken), genAffixStyle(inputToken)];
}, initComponentToken, { resetFont: false });
var style_default = genStyleHooks(["Input", "Component"], (token) => {
	const inputToken = mergeToken(token, initInputToken(token));
	return [
		genGroupStyle(inputToken),
		genRangeStyle(inputToken),
		genCompactItemStyle(inputToken, {
			focus: true,
			focusElCls: `${inputToken.componentCls}-affix-wrapper-focused`
		})
	];
}, initComponentToken, { resetFont: false });

//#endregion
export { style_default as default, genActiveStyle, genAffixStyle, genBasicInputStyle, genInputGroupStyle, genInputLargeStyle, genInputSmallStyle, genInputStyle, genPlaceholderStyle, initComponentToken, initInputToken, useSharedStyle };
