// antdv-next@1.2.2/dist/input/style/otp.js
import { genStyleHooks } from "../../theme/util/genStyleUtils.js";
import { mergeToken } from "../../theme/internal.js";
import { initComponentToken, initInputToken } from "./token.js";

//#region src/input/style/otp.ts
const genOTPStyle = (token) => {
	const { componentCls, paddingXS } = token;
	return { [componentCls]: {
		display: "inline-flex",
		alignItems: "center",
		flexWrap: "nowrap",
		columnGap: paddingXS,
		[`${componentCls}-input-wrapper`]: {
			position: "relative",
			[`${componentCls}-mask-icon`]: {
				position: "absolute",
				zIndex: "1",
				top: "50%",
				right: "50%",
				transform: "translate(50%, -50%)",
				pointerEvents: "none"
			},
			[`${componentCls}-mask-input`]: {
				color: "transparent",
				caretColor: token.colorText,
				"&::selection": { color: "transparent" }
			},
			[`${componentCls}-mask-input[type=number]::-webkit-inner-spin-button`]: {
				"-webkit-appearance": "none",
				margin: 0
			},
			[`${componentCls}-mask-input[type=number]`]: { "-moz-appearance": "textfield" }
		},
		"&-rtl": { direction: "rtl" },
		[`${componentCls}-input`]: {
			textAlign: "center",
			paddingInline: token.paddingXXS
		},
		[`&${componentCls}-sm ${componentCls}-input`]: { paddingInline: token.calc(token.paddingXXS).div(2).equal() },
		[`&${componentCls}-lg ${componentCls}-input`]: { paddingInline: token.paddingXS }
	} };
};
var otp_default = genStyleHooks(["Input", "OTP"], (token) => {
	return genOTPStyle(mergeToken(token, initInputToken(token)));
}, initComponentToken);

//#endregion
export { otp_default as default };
