// antdv-next@1.2.2/dist/input-number/style/index.js
import { resetComponent, resetIcon } from "../../style/index.js";
import { genCssVar, genStyleHooks } from "../../theme/util/genStyleUtils.js";
import { mergeToken } from "../../theme/internal.js";
import { genCompactItemStyle } from "../../style/compact-item.js";
import { initInputToken } from "../../input/style/token.js";
import { genBorderlessStyle, genFilledStyle, genOutlinedStyle, genUnderlinedStyle } from "../../input/style/variants.js";
import { genBasicInputStyle, genPlaceholderStyle } from "../../input/style/index.js";
import { prepareComponentToken } from "./token.js";
import { unit } from "@antdv-next/cssinjs";

//#region src/input-number/style/index.ts
const genInputNumberStyles = (token) => {
	const { componentCls, lineWidth, lineType, borderRadius, inputFontSizeSM, inputFontSizeLG, colorError, paddingInlineSM, paddingBlockSM, paddingBlockLG, paddingInlineLG, colorIcon, motionDurationMid, handleHoverColor, handleOpacity, paddingInline, paddingBlock, handleBg, handleActiveBg, inputAffixPadding, borderRadiusSM, controlWidth, handleBorderColor, filledHandleBg, lineHeightLG, antCls, colorTextDisabled } = token;
	const borderStyle = `${unit(lineWidth)} ${lineType} ${handleBorderColor}`;
	const [varName, varRef] = genCssVar(antCls, "input-number");
	return [
		{ [componentCls]: {
			...resetComponent(token),
			...genBasicInputStyle(token),
			[varName("input-padding-block")]: unit(paddingBlock),
			[varName("input-padding-inline")]: unit(paddingInline),
			display: "inline-flex",
			width: controlWidth,
			margin: 0,
			paddingBlock: 0,
			borderRadius,
			...genOutlinedStyle(token, { [`${componentCls}-actions`]: {
				background: handleBg,
				[`${componentCls}-action-down`]: { borderBlockStart: borderStyle }
			} }),
			...genFilledStyle(token, {
				[`${componentCls}-actions`]: {
					background: filledHandleBg,
					[`${componentCls}-action-down`]: { borderBlockStart: borderStyle }
				},
				"&:focus-within": { [`${componentCls}-actions`]: { background: handleBg } }
			}),
			...genUnderlinedStyle(token, { [`${componentCls}-actions`]: {
				background: handleBg,
				[`${componentCls}-action-down`]: { borderBlockStart: borderStyle }
			} }),
			...genBorderlessStyle(token),
			[`&${componentCls}-borderless`]: {
				paddingBlock: 0,
				[varName("input-padding-block")]: unit(token.calc(paddingBlock).add(lineWidth).equal())
			},
			[`&${componentCls}-borderless${componentCls}-sm`]: {
				paddingBlock: 0,
				[varName("input-padding-block")]: unit(token.calc(paddingBlockSM).add(lineWidth).equal())
			},
			[`&${componentCls}-borderless${componentCls}-lg`]: {
				paddingBlock: 0,
				[varName("input-padding-block")]: unit(token.calc(paddingBlockLG).add(lineWidth).equal())
			},
			"&-rtl": {
				direction: "rtl",
				[`${componentCls}-input`]: { direction: "rtl" }
			},
			[`&${componentCls}-out-of-range`]: { [`${componentCls}-input`]: { color: colorError } },
			[`${componentCls}-input`]: {
				...resetComponent(token),
				width: "100%",
				paddingBlock: varRef("input-padding-block"),
				textAlign: "start",
				backgroundColor: "transparent",
				border: 0,
				borderRadius: 0,
				outline: 0,
				transition: `all ${motionDurationMid} linear`,
				appearance: "textfield",
				fontSize: "inherit",
				lineHeight: "inherit",
				...genPlaceholderStyle(token.colorTextPlaceholder),
				"&[type=\"number\"]::-webkit-inner-spin-button, &[type=\"number\"]::-webkit-outer-spin-button": {
					margin: 0,
					appearance: "none"
				}
			},
			[`&:hover ${componentCls}-handler-wrap, &-focused ${componentCls}-handler-wrap`]: {
				width: token.handleWidth,
				opacity: 1
			},
			[`&-disabled ${componentCls}-input`]: {
				cursor: "not-allowed",
				color: token.colorTextDisabled
			}
		} },
		{ [componentCls]: {
			[`${componentCls}-action`]: {
				...resetIcon(),
				userSelect: "none",
				overflow: "hidden",
				fontWeight: "bold",
				lineHeight: 0,
				textAlign: "center",
				cursor: "pointer",
				transition: `all ${motionDurationMid} linear`,
				[`&:active:not(${componentCls}-action-up-disabled):not(${componentCls}-action-down-disabled)`]: { background: handleActiveBg },
				[`&:hover:not(${componentCls}-action-up-disabled):not(${componentCls}-action-down-disabled)`]: { color: handleHoverColor },
				[`&${componentCls}-action-up-disabled, &${componentCls}-action-down-disabled`]: {
					cursor: "not-allowed",
					color: colorTextDisabled
				}
			},
			"&-mode-input": {
				overflow: "hidden",
				[`${componentCls}-actions`]: {
					position: "absolute",
					insetBlockStart: 0,
					insetInlineEnd: 0,
					width: token.handleVisibleWidth,
					opacity: handleOpacity,
					height: "100%",
					borderRadius: 0,
					display: "flex",
					flexDirection: "column",
					alignItems: "stretch",
					transition: `all ${motionDurationMid}`,
					overflow: "hidden",
					[`${componentCls}-action`]: {
						display: "flex",
						alignItems: "center",
						justifyContent: "center",
						flex: "auto",
						height: "40%",
						marginInlineEnd: 0,
						fontSize: token.handleFontSize
					}
				},
				[`&:hover ${componentCls}-actions, &-focused ${componentCls}-actions`]: {
					width: token.handleWidth,
					opacity: 1
				},
				[`${componentCls}-action`]: {
					color: colorIcon,
					height: "50%",
					borderInlineStart: borderStyle,
					[`&:hover:not(${componentCls}-action-up-disabled):not(${componentCls}-action-down-disabled)`]: { height: `60%` }
				},
				[`&${componentCls}-disabled, &${componentCls}-readonly`]: { [`${componentCls}-actions`]: { display: "none" } }
			},
			[`&${componentCls}-mode-spinner`]: {
				padding: 0,
				width: "auto",
				[`${componentCls}-action`]: {
					flex: "none",
					paddingInline: varRef("input-padding-inline"),
					"&-up": { borderInlineStart: borderStyle },
					"&-down": { borderInlineEnd: borderStyle }
				},
				[`${componentCls}-input`]: {
					textAlign: "center",
					paddingInline: varRef("input-padding-inline")
				}
			}
		} },
		{ [componentCls]: {
			"&-lg": {
				[varName("input-padding-block")]: unit(paddingBlockLG),
				[varName("input-padding-inline")]: unit(paddingInlineLG),
				paddingBlock: 0,
				fontSize: inputFontSizeLG,
				lineHeight: lineHeightLG
			},
			"&-sm": {
				[varName("input-padding-block")]: unit(paddingBlockSM),
				[varName("input-padding-inline")]: unit(paddingInlineSM),
				paddingBlock: 0,
				fontSize: inputFontSizeSM,
				borderRadius: borderRadiusSM
			}
		} },
		{ [componentCls]: {
			[`${componentCls}-prefix, ${componentCls}-suffix`]: {
				display: "flex",
				flex: "none",
				alignItems: "center",
				alignSelf: "center",
				pointerEvents: "none"
			},
			[`${componentCls}-prefix`]: { marginInlineEnd: inputAffixPadding },
			[`${componentCls}-suffix`]: {
				height: "100%",
				marginInlineStart: inputAffixPadding,
				transition: `margin ${motionDurationMid}`
			},
			[`&:hover:not(${componentCls}-without-controls)`]: { [`${componentCls}-suffix`]: { marginInlineEnd: token.handleWidth } }
		} }
	];
};
const genCompatibleStyles = (token) => {
	const { componentCls, antCls } = token;
	return { [`${componentCls}-addon`]: { [`&:has(${antCls}-select)`]: {
		border: 0,
		padding: 0
	} } };
};
var style_default = genStyleHooks("InputNumber", (token) => {
	const inputNumberToken = mergeToken(token, initInputToken(token));
	return [
		genInputNumberStyles(inputNumberToken),
		genCompatibleStyles(inputNumberToken),
		genCompactItemStyle(inputNumberToken)
	];
}, prepareComponentToken, {
	unitless: { handleOpacity: true },
	resetFont: false
});

//#endregion
export { style_default as default };
