// antdv-next@1.2.2/dist/cascader/style/columns.js
import { textEllipsis } from "../../style/index.js";
import { getStyle } from "../../checkbox/style/index.js";
import { unit } from "@antdv-next/cssinjs";

//#region src/cascader/style/columns.ts
const getColumnsStyle = (token) => {
	const { prefixCls, componentCls } = token;
	const cascaderMenuItemCls = `${componentCls}-menu-item`;
	const iconCls = `
  &${cascaderMenuItemCls}-expand ${cascaderMenuItemCls}-expand-icon,
  ${cascaderMenuItemCls}-loading-icon
`;
	return [getStyle(`${prefixCls}-checkbox`, token), { [componentCls]: {
		"&-checkbox": {
			top: 0,
			marginInlineEnd: token.paddingXS,
			pointerEvents: "unset"
		},
		"&-menus": {
			display: "flex",
			flexWrap: "nowrap",
			alignItems: "flex-start",
			[`&${componentCls}-menu-empty`]: { [`${componentCls}-menu`]: {
				width: "100%",
				height: "auto",
				[cascaderMenuItemCls]: { color: token.colorTextDisabled }
			} }
		},
		"&-menu": {
			flexGrow: 1,
			flexShrink: 0,
			minWidth: token.controlItemWidth,
			height: token.dropdownHeight,
			margin: 0,
			padding: token.menuPadding,
			overflow: "auto",
			verticalAlign: "top",
			listStyle: "none",
			"-ms-overflow-style": "-ms-autohiding-scrollbar",
			"&:not(:last-child)": { borderInlineEnd: `${unit(token.lineWidth)} ${token.lineType} ${token.colorSplit}` },
			"&-item": {
				display: "flex",
				maxWidth: 400,
				flexWrap: "nowrap",
				alignItems: "center",
				padding: token.optionPadding,
				lineHeight: token.lineHeight,
				cursor: "pointer",
				transition: `all ${token.motionDurationMid}`,
				borderRadius: token.borderRadiusSM,
				"&:hover": { background: token.controlItemBgHover },
				"&-disabled": {
					color: token.colorTextDisabled,
					cursor: "not-allowed",
					"&:hover": { background: "transparent" },
					[iconCls]: { color: token.colorTextDisabled }
				},
				[`&-active:not(${cascaderMenuItemCls}-disabled)`]: { "&, &:hover": {
					color: token.optionSelectedColor,
					fontWeight: token.optionSelectedFontWeight,
					backgroundColor: token.optionSelectedBg
				} },
				"&-content": {
					flex: "auto",
					minWidth: 0,
					...textEllipsis
				},
				[iconCls]: {
					marginInlineStart: token.paddingXXS,
					color: token.colorIcon,
					fontSize: token.fontSizeIcon
				},
				"&-keyword": { color: token.colorHighlight }
			}
		}
	} }];
};
var columns_default = getColumnsStyle;

//#endregion
export { columns_default as default };
