// antdv-next@1.2.2/dist/theme/util/alias.js
import seed_default from "../themes/seed.js";
import getAlphaColor_default from "./getAlphaColor.js";
import { FastColor } from "@ant-design/fast-color";

//#region src/theme/util/alias.ts
/**
* Seed (designer) > Derivative (designer) > Alias (developer).
*
* Merge seed & derivative & override token and generate alias token for developer.
*/
function formatToken(derivativeToken) {
	const { override, ...restToken } = derivativeToken;
	const overrideTokens = { ...override };
	Object.keys(seed_default).forEach((token) => {
		delete overrideTokens[token];
	});
	const mergedToken = {
		...restToken,
		...overrideTokens
	};
	const shadowBaseColor = new FastColor(mergedToken.colorShadow);
	const shadowBaseAlpha = shadowBaseColor.a;
	const getShadowColor = (alpha) => shadowBaseColor.clone().setA(shadowBaseAlpha * alpha).toRgbString();
	const screenXS = 480;
	const screenSM = 576;
	const screenMD = 768;
	const screenLG = 992;
	const screenXL = 1200;
	const screenXXL = 1600;
	const screenXXXL = 1920;
	if (mergedToken.motion === false) {
		const fastDuration = "0s";
		mergedToken.motionDurationFast = fastDuration;
		mergedToken.motionDurationMid = fastDuration;
		mergedToken.motionDurationSlow = fastDuration;
	}
	return {
		...mergedToken,
		colorFillContent: mergedToken.colorFillSecondary,
		colorFillContentHover: mergedToken.colorFill,
		colorFillAlter: mergedToken.colorFillQuaternary,
		colorBgContainerDisabled: mergedToken.colorFillTertiary,
		colorBorderBg: mergedToken.colorBgContainer,
		colorSplit: getAlphaColor_default(mergedToken.colorBorderSecondary, mergedToken.colorBgContainer),
		colorTextPlaceholder: mergedToken.colorTextQuaternary,
		colorTextDisabled: mergedToken.colorTextQuaternary,
		colorTextHeading: mergedToken.colorText,
		colorTextLabel: mergedToken.colorTextSecondary,
		colorTextDescription: mergedToken.colorTextTertiary,
		colorTextLightSolid: mergedToken.colorWhite,
		colorHighlight: mergedToken.colorError,
		colorBgTextHover: mergedToken.colorFillSecondary,
		colorBgTextActive: mergedToken.colorFill,
		colorIcon: mergedToken.colorTextTertiary,
		colorIconHover: mergedToken.colorText,
		colorErrorOutline: getAlphaColor_default(mergedToken.colorErrorBg, mergedToken.colorBgContainer),
		colorWarningOutline: getAlphaColor_default(mergedToken.colorWarningBg, mergedToken.colorBgContainer),
		fontSizeIcon: mergedToken.fontSizeSM,
		lineWidthFocus: mergedToken.lineWidth * 3,
		lineWidth: mergedToken.lineWidth,
		controlOutlineWidth: mergedToken.lineWidth * 2,
		controlInteractiveSize: mergedToken.controlHeight / 2,
		controlItemBgHover: mergedToken.colorFillTertiary,
		controlItemBgActive: mergedToken.colorPrimaryBg,
		controlItemBgActiveHover: mergedToken.colorPrimaryBgHover,
		controlItemBgActiveDisabled: mergedToken.colorFill,
		controlTmpOutline: mergedToken.colorFillQuaternary,
		controlOutline: getAlphaColor_default(mergedToken.colorPrimaryBg, mergedToken.colorBgContainer),
		lineType: mergedToken.lineType,
		borderRadius: mergedToken.borderRadius,
		borderRadiusXS: mergedToken.borderRadiusXS,
		borderRadiusSM: mergedToken.borderRadiusSM,
		borderRadiusLG: mergedToken.borderRadiusLG,
		fontWeightStrong: 600,
		opacityLoading: .65,
		linkDecoration: "none",
		linkHoverDecoration: "none",
		linkFocusDecoration: "none",
		controlPaddingHorizontal: 12,
		controlPaddingHorizontalSM: 8,
		paddingXXS: mergedToken.sizeXXS,
		paddingXS: mergedToken.sizeXS,
		paddingSM: mergedToken.sizeSM,
		padding: mergedToken.size,
		paddingMD: mergedToken.sizeMD,
		paddingLG: mergedToken.sizeLG,
		paddingXL: mergedToken.sizeXL,
		paddingContentHorizontalLG: mergedToken.sizeLG,
		paddingContentVerticalLG: mergedToken.sizeMS,
		paddingContentHorizontal: mergedToken.sizeMS,
		paddingContentVertical: mergedToken.sizeSM,
		paddingContentHorizontalSM: mergedToken.size,
		paddingContentVerticalSM: mergedToken.sizeXS,
		marginXXS: mergedToken.sizeXXS,
		marginXS: mergedToken.sizeXS,
		marginSM: mergedToken.sizeSM,
		margin: mergedToken.size,
		marginMD: mergedToken.sizeMD,
		marginLG: mergedToken.sizeLG,
		marginXL: mergedToken.sizeXL,
		marginXXL: mergedToken.sizeXXL,
		boxShadow: `
      0 6px 16px 0 ${getShadowColor(.08)},
      0 3px 6px -4px ${getShadowColor(.12)},
      0 9px 28px 8px ${getShadowColor(.05)}
    `,
		boxShadowSecondary: `
      0 6px 16px 0 ${getShadowColor(.08)},
      0 3px 6px -4px ${getShadowColor(.12)},
      0 9px 28px 8px ${getShadowColor(.05)}
    `,
		boxShadowTertiary: `
      0 1px 2px 0 ${getShadowColor(.03)},
      0 1px 6px -1px ${getShadowColor(.02)},
      0 2px 4px 0 ${getShadowColor(.02)}
    `,
		screenXS,
		screenXSMin: screenXS,
		screenXSMax: screenSM - 1,
		screenSM,
		screenSMMin: screenSM,
		screenSMMax: screenMD - 1,
		screenMD,
		screenMDMin: screenMD,
		screenMDMax: screenLG - 1,
		screenLG,
		screenLGMin: screenLG,
		screenLGMax: screenXL - 1,
		screenXL,
		screenXLMin: screenXL,
		screenXLMax: screenXXL - 1,
		screenXXL,
		screenXXLMin: screenXXL,
		screenXXLMax: screenXXXL - 1,
		screenXXXL,
		screenXXXLMin: screenXXXL,
		boxShadowPopoverArrow: `2px 2px 5px ${getShadowColor(.05)}`,
		boxShadowCard: `
      0 1px 2px -2px ${getShadowColor(.16)},
      0 3px 6px 0 ${getShadowColor(.12)},
      0 5px 12px 4px ${getShadowColor(.09)}
    `,
		boxShadowDrawerRight: `
      -6px 0 16px 0 ${getShadowColor(.08)},
      -3px 0 6px -4px ${getShadowColor(.12)},
      -9px 0 28px 8px ${getShadowColor(.05)}
    `,
		boxShadowDrawerLeft: `
      6px 0 16px 0 ${getShadowColor(.08)},
      3px 0 6px -4px ${getShadowColor(.12)},
      9px 0 28px 8px ${getShadowColor(.05)}
    `,
		boxShadowDrawerUp: `
      0 6px 16px 0 ${getShadowColor(.08)},
      0 3px 6px -4px ${getShadowColor(.12)},
      0 9px 28px 8px ${getShadowColor(.05)}
    `,
		boxShadowDrawerDown: `
      0 -6px 16px 0 ${getShadowColor(.08)},
      0 -3px 6px -4px ${getShadowColor(.12)},
      0 -9px 28px 8px ${getShadowColor(.05)}
    `,
		boxShadowTabsOverflowLeft: `inset 10px 0 8px -8px ${getShadowColor(.08)}`,
		boxShadowTabsOverflowRight: `inset -10px 0 8px -8px ${getShadowColor(.08)}`,
		boxShadowTabsOverflowTop: `inset 0 10px 8px -8px ${getShadowColor(.08)}`,
		boxShadowTabsOverflowBottom: `inset 0 -10px 8px -8px ${getShadowColor(.08)}`,
		...overrideTokens
	};
}

//#endregion
export { formatToken as default };
