// @v-c/slick@1.0.2/dist/default-props.js
import { createVNode } from "vue";
var default_props_default = {
	accessibility: true,
	adaptiveHeight: false,
	afterChange: null,
	appendDots: (dots) => createVNode("ul", { "style": { display: "block" } }, [dots]),
	arrows: true,
	autoplay: false,
	autoplaySpeed: 3e3,
	beforeChange: null,
	centerMode: false,
	centerPadding: "50px",
	className: "",
	cssEase: "ease",
	customPaging: (i) => createVNode("button", null, [i + 1]),
	dots: false,
	dotsClass: "slick-dots",
	draggable: true,
	easing: "linear",
	edgeFriction: .35,
	fade: false,
	focusOnSelect: false,
	infinite: true,
	initialSlide: 0,
	lazyLoad: null,
	nextArrow: null,
	onEdge: null,
	onInit: null,
	onLazyLoadError: null,
	onReInit: null,
	pauseOnDotsHover: false,
	pauseOnFocus: false,
	pauseOnHover: true,
	prevArrow: null,
	responsive: null,
	rows: 1,
	rtl: false,
	slide: "div",
	slidesPerRow: 1,
	slidesToScroll: 1,
	slidesToShow: 1,
	speed: 500,
	swipe: true,
	swipeEvent: null,
	swipeToSlide: false,
	touchMove: true,
	touchThreshold: 5,
	useCSS: true,
	useTransform: true,
	variableWidth: false,
	vertical: false,
	verticalSwiping: false,
	waitForAnimate: true,
	asNavFor: null,
	unslick: false
};
export { default_props_default as default };
