// antdv-next@1.2.2/dist/radio/style/index.js
import { genFocusOutline, resetComponent } from "../../style/index.js";
import { genStyleHooks } from "../../theme/util/genStyleUtils.js";
import { mergeToken } from "../../theme/internal.js";
import { genNoMotionStyle } from "../../style/motion/util.js";
import { unit } from "@antdv-next/cssinjs";

//#region src/radio/style/index.ts
const getGroupRadioStyle = (token) => {
	const { componentCls, antCls } = token;
	const groupPrefixCls = `${componentCls}-group`;
	return { [groupPrefixCls]: {
		...resetComponent(token),
		display: "inline-block",
		fontSize: 0,
		[`&${groupPrefixCls}-rtl`]: { direction: "rtl" },
		[`&${groupPrefixCls}-block`]: { display: "flex" },
		[`${antCls}-badge ${antCls}-badge-count`]: { zIndex: 1 },
		[`> ${antCls}-badge:not(:first-child) > ${antCls}-button-wrapper`]: { borderInlineStart: "none" },
		"&-vertical": {
			display: "flex",
			flexDirection: "column",
			rowGap: token.marginXS,
			[`${componentCls}-wrapper`]: { marginInlineEnd: 0 }
		}
	} };
};
const getRadioBasicStyle = (token) => {
	const { componentCls, wrapperMarginInlineEnd, colorPrimary, colorPrimaryHover, radioSize, motionDurationSlow, motionDurationMid, motionEaseInOutCirc, colorBgContainer, colorBorder, lineWidth, colorBgContainerDisabled, colorTextDisabled, paddingXS, dotColorDisabled, dotSize, lineType, radioColor, radioBgColor } = token;
	return { [`${componentCls}-wrapper`]: {
		...resetComponent(token),
		display: "inline-flex",
		alignItems: "baseline",
		marginInlineStart: 0,
		marginInlineEnd: wrapperMarginInlineEnd,
		cursor: "pointer",
		"&:last-child": { marginInlineEnd: 0 },
		[`&${componentCls}-wrapper-rtl`]: { direction: "rtl" },
		"&-disabled": {
			cursor: "not-allowed",
			color: token.colorTextDisabled
		},
		"&::after": {
			display: "inline-block",
			width: 0,
			overflow: "hidden",
			content: "\"\\a0\""
		},
		"&-block": {
			flex: 1,
			justifyContent: "center"
		},
		[componentCls]: {
			...resetComponent(token),
			position: "relative",
			whiteSpace: "nowrap",
			lineHeight: 1,
			cursor: "pointer",
			alignSelf: "center",
			boxSizing: "border-box",
			display: "block",
			flex: "none",
			width: `calc(${radioSize} * 1px)`,
			height: `calc(${radioSize} * 1px)`,
			backgroundColor: colorBgContainer,
			border: `${unit(lineWidth)} ${lineType} ${colorBorder}`,
			borderRadius: "50%",
			transition: `all ${motionDurationMid}`,
			...genNoMotionStyle(),
			"&:after": {
				content: "\"\"",
				position: "absolute",
				top: "50%",
				left: "50%",
				transform: "translate(-50%, -50%) scale(0)",
				width: `calc(${dotSize} * 1px)`,
				height: `calc(${dotSize} * 1px)`,
				backgroundColor: radioColor,
				borderRadius: "50%",
				transformOrigin: "50% 50%",
				opacity: 0,
				transition: `all ${motionDurationSlow} ${motionEaseInOutCirc}`,
				...genNoMotionStyle()
			},
			[`${componentCls}-input`]: {
				position: "absolute",
				inset: 0,
				zIndex: 1,
				cursor: "pointer",
				opacity: 0,
				margin: 0
			},
			[`&:has(${componentCls}-input:focus-visible)`]: genFocusOutline(token)
		},
		[`&:hover:not(${componentCls}-wrapper-disabled) ${componentCls}`]: { borderColor: colorPrimary },
		[`&:hover ${componentCls}-checked:not(${componentCls}-disabled)`]: {
			backgroundColor: colorPrimaryHover,
			borderColor: "transparent"
		},
		[`${componentCls}-checked`]: {
			backgroundColor: radioBgColor,
			borderColor: colorPrimary,
			"&::after": {
				transform: `translate(-50%, -50%)`,
				opacity: 1
			}
		},
		[`${componentCls}-disabled`]: {
			[`&, ${componentCls}-input`]: {
				cursor: "not-allowed",
				pointerEvents: "none"
			},
			background: colorBgContainerDisabled,
			borderColor: colorBorder,
			"&::after": { backgroundColor: dotColorDisabled }
		},
		[`${componentCls}-disabled + span`]: {
			color: colorTextDisabled,
			cursor: "not-allowed"
		},
		[`span${componentCls} + *`]: {
			paddingInlineStart: paddingXS,
			paddingInlineEnd: paddingXS
		}
	} };
};
const getRadioButtonStyle = (token) => {
	const { buttonColor, controlHeight, componentCls, lineWidth, lineType, colorBorder, motionDurationMid, buttonPaddingInline, fontSize, buttonBg, fontSizeLG, controlHeightLG, controlHeightSM, paddingXS, borderRadius, borderRadiusSM, borderRadiusLG, buttonCheckedBg, buttonSolidCheckedColor, colorTextDisabled, colorBgContainerDisabled, buttonCheckedBgDisabled, buttonCheckedColorDisabled, colorPrimary, colorPrimaryHover, colorPrimaryActive, buttonSolidCheckedBg, buttonSolidCheckedHoverBg, buttonSolidCheckedActiveBg, calc } = token;
	return { [`${componentCls}-button-wrapper`]: {
		position: "relative",
		display: "inline-block",
		height: controlHeight,
		margin: 0,
		paddingInline: buttonPaddingInline,
		paddingBlock: 0,
		color: buttonColor,
		fontSize,
		lineHeight: unit(calc(controlHeight).sub(calc(lineWidth).mul(2)).equal()),
		background: buttonBg,
		border: `${unit(lineWidth)} ${lineType} ${colorBorder}`,
		borderBlockStartWidth: calc(lineWidth).add(.02).equal(),
		borderInlineEndWidth: lineWidth,
		cursor: "pointer",
		transition: [
			`color`,
			`background-color`,
			`box-shadow`
		].map((prop) => `${prop} ${motionDurationMid}`).join(","),
		...genNoMotionStyle(),
		a: { color: buttonColor },
		[`> ${componentCls}-button`]: {
			position: "absolute",
			insetBlockStart: 0,
			insetInlineStart: 0,
			zIndex: -1,
			width: "100%",
			height: "100%"
		},
		"&:not(:last-child)": { marginInlineEnd: calc(lineWidth).mul(-1).equal() },
		"&:first-child": {
			borderInlineStart: `${unit(lineWidth)} ${lineType} ${colorBorder}`,
			borderStartStartRadius: borderRadius,
			borderEndStartRadius: borderRadius
		},
		"&:last-child": {
			borderStartEndRadius: borderRadius,
			borderEndEndRadius: borderRadius
		},
		"&:first-child:last-child": { borderRadius },
		[`${componentCls}-group-large &`]: {
			height: controlHeightLG,
			fontSize: fontSizeLG,
			lineHeight: unit(calc(controlHeightLG).sub(calc(lineWidth).mul(2)).equal()),
			"&:first-child": {
				borderStartStartRadius: borderRadiusLG,
				borderEndStartRadius: borderRadiusLG
			},
			"&:last-child": {
				borderStartEndRadius: borderRadiusLG,
				borderEndEndRadius: borderRadiusLG
			}
		},
		[`${componentCls}-group-small &`]: {
			height: controlHeightSM,
			paddingInline: calc(paddingXS).sub(lineWidth).equal(),
			paddingBlock: 0,
			lineHeight: unit(calc(controlHeightSM).sub(calc(lineWidth).mul(2)).equal()),
			"&:first-child": {
				borderStartStartRadius: borderRadiusSM,
				borderEndStartRadius: borderRadiusSM
			},
			"&:last-child": {
				borderStartEndRadius: borderRadiusSM,
				borderEndEndRadius: borderRadiusSM
			}
		},
		"&:hover": {
			position: "relative",
			color: colorPrimary
		},
		"&:has(:focus-visible)": genFocusOutline(token),
		[`${componentCls}, input[type='checkbox'], input[type='radio']`]: {
			width: 0,
			height: 0,
			opacity: 0,
			pointerEvents: "none"
		},
		[`&-checked:not(${componentCls}-button-wrapper-disabled)`]: {
			zIndex: 1,
			color: colorPrimary,
			background: buttonCheckedBg,
			borderColor: colorPrimary,
			"&::before": { backgroundColor: colorPrimary },
			"&:first-child": { borderColor: colorPrimary },
			"&:hover": {
				color: colorPrimaryHover,
				borderColor: colorPrimaryHover,
				"&::before": { backgroundColor: colorPrimaryHover }
			},
			"&:active": {
				color: colorPrimaryActive,
				borderColor: colorPrimaryActive,
				"&::before": { backgroundColor: colorPrimaryActive }
			}
		},
		[`${componentCls}-group-solid &-checked:not(${componentCls}-button-wrapper-disabled)`]: {
			color: buttonSolidCheckedColor,
			background: buttonSolidCheckedBg,
			borderColor: buttonSolidCheckedBg,
			"&:hover": {
				color: buttonSolidCheckedColor,
				background: buttonSolidCheckedHoverBg,
				borderColor: buttonSolidCheckedHoverBg
			},
			"&:active": {
				color: buttonSolidCheckedColor,
				background: buttonSolidCheckedActiveBg,
				borderColor: buttonSolidCheckedActiveBg
			}
		},
		"&-disabled": {
			color: colorTextDisabled,
			backgroundColor: colorBgContainerDisabled,
			borderColor: colorBorder,
			cursor: "not-allowed",
			"&:first-child, &:hover": {
				color: colorTextDisabled,
				backgroundColor: colorBgContainerDisabled,
				borderColor: colorBorder
			}
		},
		[`&-disabled${componentCls}-button-wrapper-checked`]: {
			color: buttonCheckedColorDisabled,
			backgroundColor: buttonCheckedBgDisabled,
			borderColor: colorBorder,
			boxShadow: "none"
		},
		"&-block": {
			flex: 1,
			textAlign: "center"
		}
	} };
};
const prepareComponentToken = (token) => {
	const { wireframe, padding, marginXS, lineWidth, fontSizeLG, colorText, colorBgContainer, colorTextDisabled, controlItemBgActiveDisabled, colorTextLightSolid, colorPrimary, colorPrimaryHover, colorPrimaryActive, colorWhite } = token;
	const dotPadding = 4;
	const radioSize = fontSizeLG;
	return {
		radioSize,
		dotSize: wireframe ? radioSize - dotPadding * 2 : radioSize - (dotPadding + lineWidth) * 2,
		dotColorDisabled: colorTextDisabled,
		buttonSolidCheckedColor: colorTextLightSolid,
		buttonSolidCheckedBg: colorPrimary,
		buttonSolidCheckedHoverBg: colorPrimaryHover,
		buttonSolidCheckedActiveBg: colorPrimaryActive,
		buttonBg: colorBgContainer,
		buttonCheckedBg: colorBgContainer,
		buttonColor: colorText,
		buttonCheckedBgDisabled: controlItemBgActiveDisabled,
		buttonCheckedColorDisabled: colorTextDisabled,
		buttonPaddingInline: padding - lineWidth,
		wrapperMarginInlineEnd: marginXS,
		radioColor: wireframe ? colorPrimary : colorWhite,
		radioBgColor: wireframe ? colorBgContainer : colorPrimary
	};
};
var style_default = genStyleHooks("Radio", (token) => {
	const { controlOutline, controlOutlineWidth } = token;
	const radioFocusShadow = `0 0 0 ${unit(controlOutlineWidth)} ${controlOutline}`;
	const radioToken = mergeToken(token, {
		radioFocusShadow,
		radioButtonFocusShadow: radioFocusShadow
	});
	return [
		getGroupRadioStyle(radioToken),
		getRadioBasicStyle(radioToken),
		getRadioButtonStyle(radioToken)
	];
}, prepareComponentToken, { unitless: {
	radioSize: true,
	dotSize: true
} });

//#endregion
export { style_default as default, prepareComponentToken };
