// @antdv-next/auto-import-resolver@1.0.7/dist/index.mjs
//#region src/components.ts
var components_default = {
	AAffix: "Affix",
	AButton: "Button",
	AAlert: "Alert",
	AAnchor: "Anchor",
	ASpace: "Space",
	ADivider: "Divider",
	ASpaceCompact: "SpaceCompact",
	ASpaceAddon: "SpaceAddon",
	AStyleProvider: "StyleProvider",
	AFlex: "Flex",
	AConfigProvider: "ConfigProvider",
	ACol: "Col",
	ARow: "Row",
	ABadge: "Badge",
	ABadgeRibbon: "BadgeRibbon",
	ATag: "Tag",
	ACheckableTag: "CheckableTag",
	ACheckableTagGroup: "CheckableTagGroup",
	ALayout: "Layout",
	ALayoutHeader: "LayoutHeader",
	ALayoutFooter: "LayoutFooter",
	ALayoutContent: "LayoutContent",
	ALayoutSider: "LayoutSider",
	AEmpty: "Empty",
	AResult: "Result",
	ASpin: "Spin",
	ASkeleton: "Skeleton",
	ASkeletonButton: "SkeletonButton",
	ASkeletonAvatar: "SkeletonAvatar",
	ASkeletonInput: "SkeletonInput",
	ASkeletonImage: "SkeletonImage",
	ASkeletonNode: "SkeletonNode",
	AStatistic: "Statistic",
	AStatisticTimer: "StatisticTimer",
	ADescriptions: "Descriptions",
	ADescriptionsItem: "DescriptionsItem",
	ATimeline: "Timeline",
	ATimelineItem: "TimelineItem",
	AWatermark: "Watermark",
	AQrcode: "QRCode",
	ACollapse: "Collapse",
	ACollapsePanel: "CollapsePanel",
	ATooltip: "Tooltip",
	ASegmented: "Segmented",
	APopover: "Popover",
	APopconfirm: "Popconfirm",
	AFloatButton: "FloatButton",
	AFloatBackTop: "BackTop",
	AFloatButtonGroup: "FloatButtonGroup",
	AMenu: "Menu",
	ASubMenu: "SubMenu",
	AMenuItem: "MenuItem",
	AMenuItemGroup: "MenuItemGroup",
	AMenuDivider: "MenuDivider",
	ADropdown: "Dropdown",
	ATour: "Tour",
	ABreadcrumb: "Breadcrumb",
	ABreadcrumbItem: "BreadcrumbItem",
	ABreadcrumbSeparator: "BreadcrumbSeparator",
	AAvatar: "Avatar",
	AAvatarGroup: "AvatarGroup",
	AForm: "Form",
	AFormItem: "FormItem",
	ACheckbox: "Checkbox",
	ACheckboxGroup: "CheckboxGroup",
	ARadio: "Radio",
	ARadioGroup: "RadioGroup",
	ARadioButton: "RadioButton",
	ASwitch: "Switch",
	AColorPicker: "ColorPicker",
	ASlider: "Slider",
	AInput: "Input",
	AInputSearch: "InputSearch",
	AInputPassword: "InputPassword",
	AInputOtp: "InputOTP",
	AInputGroup: "InputGroup",
	AInputNumber: "InputNumber",
	ATextarea: "TextArea",
	ATypography: "Typography",
	ATypographyText: "TypographyText",
	ATypographyTitle: "TypographyTitle",
	ATypographyParagraph: "TypographyParagraph",
	ATypographyLink: "TypographyLink",
	ATabs: "Tabs",
	ATabPane: "TabPane",
	ASelect: "Select",
	ASelectOption: "SelectOption",
	ASelectOptGroup: "SelectOptGroup",
	ASplitter: "Splitter",
	ASplitterPanel: "SplitterPanel",
	ACard: "Card",
	ACardGrid: "CardGrid",
	ACardMeta: "CardMeta",
	AMasonry: "Masonry",
	ARate: "Rate",
	AProgress: "Progress",
	ADrawer: "Drawer",
	AModal: "Modal",
	ASteps: "Steps",
	AImage: "Image",
	AImagePreviewGroup: "ImagePreviewGroup",
	AApp: "App",
	ATree: "Tree",
	ADirectoryTree: "DirectoryTree",
	ATreeSelect: "TreeSelect",
	AUpload: "Upload",
	AUploadDragger: "UploadDragger",
	APagination: "Pagination",
	ATransfer: "Transfer",
	AMentions: "Mentions",
	ACarousel: "Carousel",
	ACascader: "Cascader",
	ATable: "Table",
	ATableColumn: "TableColumn",
	ATableSummary: "TableSummary",
	ATableSummaryRow: "TableSummaryRow",
	ATableSummaryCell: "TableSummaryCell",
	ACalendar: "Calendar",
	ADatePicker: "DatePicker",
	ARangePicker: "DateRangePicker",
	ATimePicker: "TimePicker",
	ATimeRangePicker: "TimeRangePicker",
	AAutoComplete: "AutoComplete",
	AAutoCompleteOption: "AutoCompleteOption"
};

//#endregion
//#region src/icons.ts
var icons_default = [
	"AccountBookFilled",
	"AccountBookOutlined",
	"AccountBookTwoTone",
	"AimOutlined",
	"AlertFilled",
	"AlertOutlined",
	"AlertTwoTone",
	"AlibabaOutlined",
	"AlignCenterOutlined",
	"AlignLeftOutlined",
	"AlignRightOutlined",
	"AlipayCircleFilled",
	"AlipayCircleOutlined",
	"AlipayOutlined",
	"AlipaySquareFilled",
	"AliwangwangFilled",
	"AliwangwangOutlined",
	"AliyunOutlined",
	"AmazonCircleFilled",
	"AmazonOutlined",
	"AmazonSquareFilled",
	"AndroidFilled",
	"AndroidOutlined",
	"AntCloudOutlined",
	"AntDesignOutlined",
	"ApartmentOutlined",
	"ApiFilled",
	"ApiOutlined",
	"ApiTwoTone",
	"AppleFilled",
	"AppleOutlined",
	"AppstoreAddOutlined",
	"AppstoreFilled",
	"AppstoreOutlined",
	"AppstoreTwoTone",
	"AreaChartOutlined",
	"ArrowDownOutlined",
	"ArrowLeftOutlined",
	"ArrowRightOutlined",
	"ArrowUpOutlined",
	"ArrowsAltOutlined",
	"AudioFilled",
	"AudioMutedOutlined",
	"AudioOutlined",
	"AudioTwoTone",
	"AuditOutlined",
	"BackwardFilled",
	"BackwardOutlined",
	"BaiduOutlined",
	"BankFilled",
	"BankOutlined",
	"BankTwoTone",
	"BarChartOutlined",
	"BarcodeOutlined",
	"BarsOutlined",
	"BehanceCircleFilled",
	"BehanceOutlined",
	"BehanceSquareFilled",
	"BehanceSquareOutlined",
	"BellFilled",
	"BellOutlined",
	"BellTwoTone",
	"BgColorsOutlined",
	"BilibiliFilled",
	"BilibiliOutlined",
	"BlockOutlined",
	"BoldOutlined",
	"BookFilled",
	"BookOutlined",
	"BookTwoTone",
	"BorderBottomOutlined",
	"BorderHorizontalOutlined",
	"BorderInnerOutlined",
	"BorderLeftOutlined",
	"BorderOuterOutlined",
	"BorderOutlined",
	"BorderRightOutlined",
	"BorderTopOutlined",
	"BorderVerticleOutlined",
	"BorderlessTableOutlined",
	"BoxPlotFilled",
	"BoxPlotOutlined",
	"BoxPlotTwoTone",
	"BranchesOutlined",
	"BugFilled",
	"BugOutlined",
	"BugTwoTone",
	"BuildFilled",
	"BuildOutlined",
	"BuildTwoTone",
	"BulbFilled",
	"BulbOutlined",
	"BulbTwoTone",
	"CalculatorFilled",
	"CalculatorOutlined",
	"CalculatorTwoTone",
	"CalendarFilled",
	"CalendarOutlined",
	"CalendarTwoTone",
	"CameraFilled",
	"CameraOutlined",
	"CameraTwoTone",
	"CarFilled",
	"CarOutlined",
	"CarTwoTone",
	"CaretDownFilled",
	"CaretDownOutlined",
	"CaretLeftFilled",
	"CaretLeftOutlined",
	"CaretRightFilled",
	"CaretRightOutlined",
	"CaretUpFilled",
	"CaretUpOutlined",
	"CarryOutFilled",
	"CarryOutOutlined",
	"CarryOutTwoTone",
	"CheckCircleFilled",
	"CheckCircleOutlined",
	"CheckCircleTwoTone",
	"CheckOutlined",
	"CheckSquareFilled",
	"CheckSquareOutlined",
	"CheckSquareTwoTone",
	"ChromeFilled",
	"ChromeOutlined",
	"CiCircleFilled",
	"CiCircleOutlined",
	"CiCircleTwoTone",
	"CiOutlined",
	"CiTwoTone",
	"ClearOutlined",
	"ClockCircleFilled",
	"ClockCircleOutlined",
	"ClockCircleTwoTone",
	"CloseCircleFilled",
	"CloseCircleOutlined",
	"CloseCircleTwoTone",
	"CloseOutlined",
	"CloseSquareFilled",
	"CloseSquareOutlined",
	"CloseSquareTwoTone",
	"CloudDownloadOutlined",
	"CloudFilled",
	"CloudOutlined",
	"CloudServerOutlined",
	"CloudSyncOutlined",
	"CloudTwoTone",
	"CloudUploadOutlined",
	"ClusterOutlined",
	"CodeFilled",
	"CodeOutlined",
	"CodeSandboxCircleFilled",
	"CodeSandboxOutlined",
	"CodeSandboxSquareFilled",
	"CodeTwoTone",
	"CodepenCircleFilled",
	"CodepenCircleOutlined",
	"CodepenOutlined",
	"CodepenSquareFilled",
	"CoffeeOutlined",
	"ColumnHeightOutlined",
	"ColumnWidthOutlined",
	"CommentOutlined",
	"CompassFilled",
	"CompassOutlined",
	"CompassTwoTone",
	"CompressOutlined",
	"ConsoleSqlOutlined",
	"ContactsFilled",
	"ContactsOutlined",
	"ContactsTwoTone",
	"ContainerFilled",
	"ContainerOutlined",
	"ContainerTwoTone",
	"ControlFilled",
	"ControlOutlined",
	"ControlTwoTone",
	"CopyFilled",
	"CopyOutlined",
	"CopyTwoTone",
	"CopyrightCircleFilled",
	"CopyrightCircleOutlined",
	"CopyrightCircleTwoTone",
	"CopyrightOutlined",
	"CopyrightTwoTone",
	"CreditCardFilled",
	"CreditCardOutlined",
	"CreditCardTwoTone",
	"CrownFilled",
	"CrownOutlined",
	"CrownTwoTone",
	"CustomerServiceFilled",
	"CustomerServiceOutlined",
	"CustomerServiceTwoTone",
	"DashOutlined",
	"DashboardFilled",
	"DashboardOutlined",
	"DashboardTwoTone",
	"DatabaseFilled",
	"DatabaseOutlined",
	"DatabaseTwoTone",
	"DeleteColumnOutlined",
	"DeleteFilled",
	"DeleteOutlined",
	"DeleteRowOutlined",
	"DeleteTwoTone",
	"DeliveredProcedureOutlined",
	"DeploymentUnitOutlined",
	"DesktopOutlined",
	"DiffFilled",
	"DiffOutlined",
	"DiffTwoTone",
	"DingdingOutlined",
	"DingtalkCircleFilled",
	"DingtalkOutlined",
	"DingtalkSquareFilled",
	"DisconnectOutlined",
	"DiscordFilled",
	"DiscordOutlined",
	"DislikeFilled",
	"DislikeOutlined",
	"DislikeTwoTone",
	"DockerOutlined",
	"DollarCircleFilled",
	"DollarCircleOutlined",
	"DollarCircleTwoTone",
	"DollarOutlined",
	"DollarTwoTone",
	"DotChartOutlined",
	"DotNetOutlined",
	"DoubleLeftOutlined",
	"DoubleRightOutlined",
	"DownCircleFilled",
	"DownCircleOutlined",
	"DownCircleTwoTone",
	"DownOutlined",
	"DownSquareFilled",
	"DownSquareOutlined",
	"DownSquareTwoTone",
	"DownloadOutlined",
	"DragOutlined",
	"DribbbleCircleFilled",
	"DribbbleOutlined",
	"DribbbleSquareFilled",
	"DribbbleSquareOutlined",
	"DropboxCircleFilled",
	"DropboxOutlined",
	"DropboxSquareFilled",
	"EditFilled",
	"EditOutlined",
	"EditTwoTone",
	"EllipsisOutlined",
	"EnterOutlined",
	"EnvironmentFilled",
	"EnvironmentOutlined",
	"EnvironmentTwoTone",
	"EuroCircleFilled",
	"EuroCircleOutlined",
	"EuroCircleTwoTone",
	"EuroOutlined",
	"EuroTwoTone",
	"ExceptionOutlined",
	"ExclamationCircleFilled",
	"ExclamationCircleOutlined",
	"ExclamationCircleTwoTone",
	"ExclamationOutlined",
	"ExpandAltOutlined",
	"ExpandOutlined",
	"ExperimentFilled",
	"ExperimentOutlined",
	"ExperimentTwoTone",
	"ExportOutlined",
	"EyeFilled",
	"EyeInvisibleFilled",
	"EyeInvisibleOutlined",
	"EyeInvisibleTwoTone",
	"EyeOutlined",
	"EyeTwoTone",
	"FacebookFilled",
	"FacebookOutlined",
	"FallOutlined",
	"FastBackwardFilled",
	"FastBackwardOutlined",
	"FastForwardFilled",
	"FastForwardOutlined",
	"FieldBinaryOutlined",
	"FieldNumberOutlined",
	"FieldStringOutlined",
	"FieldTimeOutlined",
	"FileAddFilled",
	"FileAddOutlined",
	"FileAddTwoTone",
	"FileDoneOutlined",
	"FileExcelFilled",
	"FileExcelOutlined",
	"FileExcelTwoTone",
	"FileExclamationFilled",
	"FileExclamationOutlined",
	"FileExclamationTwoTone",
	"FileFilled",
	"FileGifOutlined",
	"FileImageFilled",
	"FileImageOutlined",
	"FileImageTwoTone",
	"FileJpgOutlined",
	"FileMarkdownFilled",
	"FileMarkdownOutlined",
	"FileMarkdownTwoTone",
	"FileOutlined",
	"FilePdfFilled",
	"FilePdfOutlined",
	"FilePdfTwoTone",
	"FilePptFilled",
	"FilePptOutlined",
	"FilePptTwoTone",
	"FileProtectOutlined",
	"FileSearchOutlined",
	"FileSyncOutlined",
	"FileTextFilled",
	"FileTextOutlined",
	"FileTextTwoTone",
	"FileTwoTone",
	"FileUnknownFilled",
	"FileUnknownOutlined",
	"FileUnknownTwoTone",
	"FileWordFilled",
	"FileWordOutlined",
	"FileWordTwoTone",
	"FileZipFilled",
	"FileZipOutlined",
	"FileZipTwoTone",
	"FilterFilled",
	"FilterOutlined",
	"FilterTwoTone",
	"FireFilled",
	"FireOutlined",
	"FireTwoTone",
	"FlagFilled",
	"FlagOutlined",
	"FlagTwoTone",
	"FolderAddFilled",
	"FolderAddOutlined",
	"FolderAddTwoTone",
	"FolderFilled",
	"FolderOpenFilled",
	"FolderOpenOutlined",
	"FolderOpenTwoTone",
	"FolderOutlined",
	"FolderTwoTone",
	"FolderViewOutlined",
	"FontColorsOutlined",
	"FontSizeOutlined",
	"ForkOutlined",
	"FormOutlined",
	"FormatPainterFilled",
	"FormatPainterOutlined",
	"ForwardFilled",
	"ForwardOutlined",
	"FrownFilled",
	"FrownOutlined",
	"FrownTwoTone",
	"FullscreenExitOutlined",
	"FullscreenOutlined",
	"FunctionOutlined",
	"FundFilled",
	"FundOutlined",
	"FundProjectionScreenOutlined",
	"FundTwoTone",
	"FundViewOutlined",
	"FunnelPlotFilled",
	"FunnelPlotOutlined",
	"FunnelPlotTwoTone",
	"GatewayOutlined",
	"GifOutlined",
	"GiftFilled",
	"GiftOutlined",
	"GiftTwoTone",
	"GithubFilled",
	"GithubOutlined",
	"GitlabFilled",
	"GitlabOutlined",
	"GlobalOutlined",
	"GoldFilled",
	"GoldOutlined",
	"GoldTwoTone",
	"GoldenFilled",
	"GoogleCircleFilled",
	"GoogleOutlined",
	"GooglePlusCircleFilled",
	"GooglePlusOutlined",
	"GooglePlusSquareFilled",
	"GoogleSquareFilled",
	"GroupOutlined",
	"HarmonyOSOutlined",
	"HddFilled",
	"HddOutlined",
	"HddTwoTone",
	"HeartFilled",
	"HeartOutlined",
	"HeartTwoTone",
	"HeatMapOutlined",
	"HighlightFilled",
	"HighlightOutlined",
	"HighlightTwoTone",
	"HistoryOutlined",
	"HolderOutlined",
	"HomeFilled",
	"HomeOutlined",
	"HomeTwoTone",
	"HourglassFilled",
	"HourglassOutlined",
	"HourglassTwoTone",
	"Html5Filled",
	"Html5Outlined",
	"Html5TwoTone",
	"IdcardFilled",
	"IdcardOutlined",
	"IdcardTwoTone",
	"IeCircleFilled",
	"IeOutlined",
	"IeSquareFilled",
	"ImportOutlined",
	"InboxOutlined",
	"InfoCircleFilled",
	"InfoCircleOutlined",
	"InfoCircleTwoTone",
	"InfoOutlined",
	"InsertRowAboveOutlined",
	"InsertRowBelowOutlined",
	"InsertRowLeftOutlined",
	"InsertRowRightOutlined",
	"InstagramFilled",
	"InstagramOutlined",
	"InsuranceFilled",
	"InsuranceOutlined",
	"InsuranceTwoTone",
	"InteractionFilled",
	"InteractionOutlined",
	"InteractionTwoTone",
	"IssuesCloseOutlined",
	"ItalicOutlined",
	"JavaOutlined",
	"JavaScriptOutlined",
	"KeyOutlined",
	"KubernetesOutlined",
	"LaptopOutlined",
	"LayoutFilled",
	"LayoutOutlined",
	"LayoutTwoTone",
	"LeftCircleFilled",
	"LeftCircleOutlined",
	"LeftCircleTwoTone",
	"LeftOutlined",
	"LeftSquareFilled",
	"LeftSquareOutlined",
	"LeftSquareTwoTone",
	"LikeFilled",
	"LikeOutlined",
	"LikeTwoTone",
	"LineChartOutlined",
	"LineHeightOutlined",
	"LineOutlined",
	"LinkOutlined",
	"LinkedinFilled",
	"LinkedinOutlined",
	"LinuxOutlined",
	"Loading3QuartersOutlined",
	"LoadingOutlined",
	"LockFilled",
	"LockOutlined",
	"LockTwoTone",
	"LoginOutlined",
	"LogoutOutlined",
	"MacCommandFilled",
	"MacCommandOutlined",
	"MailFilled",
	"MailOutlined",
	"MailTwoTone",
	"ManOutlined",
	"MedicineBoxFilled",
	"MedicineBoxOutlined",
	"MedicineBoxTwoTone",
	"MediumCircleFilled",
	"MediumOutlined",
	"MediumSquareFilled",
	"MediumWorkmarkOutlined",
	"MehFilled",
	"MehOutlined",
	"MehTwoTone",
	"MenuFoldOutlined",
	"MenuOutlined",
	"MenuUnfoldOutlined",
	"MergeCellsOutlined",
	"MergeFilled",
	"MergeOutlined",
	"MessageFilled",
	"MessageOutlined",
	"MessageTwoTone",
	"MinusCircleFilled",
	"MinusCircleOutlined",
	"MinusCircleTwoTone",
	"MinusOutlined",
	"MinusSquareFilled",
	"MinusSquareOutlined",
	"MinusSquareTwoTone",
	"MobileFilled",
	"MobileOutlined",
	"MobileTwoTone",
	"MoneyCollectFilled",
	"MoneyCollectOutlined",
	"MoneyCollectTwoTone",
	"MonitorOutlined",
	"MoonFilled",
	"MoonOutlined",
	"MoreOutlined",
	"MutedFilled",
	"MutedOutlined",
	"NodeCollapseOutlined",
	"NodeExpandOutlined",
	"NodeIndexOutlined",
	"NotificationFilled",
	"NotificationOutlined",
	"NotificationTwoTone",
	"NumberOutlined",
	"OneToOneOutlined",
	"OpenAIFilled",
	"OpenAIOutlined",
	"OrderedListOutlined",
	"PaperClipOutlined",
	"PartitionOutlined",
	"PauseCircleFilled",
	"PauseCircleOutlined",
	"PauseCircleTwoTone",
	"PauseOutlined",
	"PayCircleFilled",
	"PayCircleOutlined",
	"PercentageOutlined",
	"PhoneFilled",
	"PhoneOutlined",
	"PhoneTwoTone",
	"PicCenterOutlined",
	"PicLeftOutlined",
	"PicRightOutlined",
	"PictureFilled",
	"PictureOutlined",
	"PictureTwoTone",
	"PieChartFilled",
	"PieChartOutlined",
	"PieChartTwoTone",
	"PinterestFilled",
	"PinterestOutlined",
	"PlayCircleFilled",
	"PlayCircleOutlined",
	"PlayCircleTwoTone",
	"PlaySquareFilled",
	"PlaySquareOutlined",
	"PlaySquareTwoTone",
	"PlusCircleFilled",
	"PlusCircleOutlined",
	"PlusCircleTwoTone",
	"PlusOutlined",
	"PlusSquareFilled",
	"PlusSquareOutlined",
	"PlusSquareTwoTone",
	"PoundCircleFilled",
	"PoundCircleOutlined",
	"PoundCircleTwoTone",
	"PoundOutlined",
	"PoweroffOutlined",
	"PrinterFilled",
	"PrinterOutlined",
	"PrinterTwoTone",
	"ProductFilled",
	"ProductOutlined",
	"ProfileFilled",
	"ProfileOutlined",
	"ProfileTwoTone",
	"ProjectFilled",
	"ProjectOutlined",
	"ProjectTwoTone",
	"PropertySafetyFilled",
	"PropertySafetyOutlined",
	"PropertySafetyTwoTone",
	"PullRequestOutlined",
	"PushpinFilled",
	"PushpinOutlined",
	"PushpinTwoTone",
	"PythonOutlined",
	"QqCircleFilled",
	"QqOutlined",
	"QqSquareFilled",
	"QrcodeOutlined",
	"QuestionCircleFilled",
	"QuestionCircleOutlined",
	"QuestionCircleTwoTone",
	"QuestionOutlined",
	"RadarChartOutlined",
	"RadiusBottomleftOutlined",
	"RadiusBottomrightOutlined",
	"RadiusSettingOutlined",
	"RadiusUpleftOutlined",
	"RadiusUprightOutlined",
	"ReadFilled",
	"ReadOutlined",
	"ReconciliationFilled",
	"ReconciliationOutlined",
	"ReconciliationTwoTone",
	"RedEnvelopeFilled",
	"RedEnvelopeOutlined",
	"RedEnvelopeTwoTone",
	"RedditCircleFilled",
	"RedditOutlined",
	"RedditSquareFilled",
	"RedoOutlined",
	"ReloadOutlined",
	"RestFilled",
	"RestOutlined",
	"RestTwoTone",
	"RetweetOutlined",
	"RightCircleFilled",
	"RightCircleOutlined",
	"RightCircleTwoTone",
	"RightOutlined",
	"RightSquareFilled",
	"RightSquareOutlined",
	"RightSquareTwoTone",
	"RiseOutlined",
	"RobotFilled",
	"RobotOutlined",
	"RocketFilled",
	"RocketOutlined",
	"RocketTwoTone",
	"RollbackOutlined",
	"RotateLeftOutlined",
	"RotateRightOutlined",
	"RubyOutlined",
	"SafetyCertificateFilled",
	"SafetyCertificateOutlined",
	"SafetyCertificateTwoTone",
	"SafetyOutlined",
	"SaveFilled",
	"SaveOutlined",
	"SaveTwoTone",
	"ScanOutlined",
	"ScheduleFilled",
	"ScheduleOutlined",
	"ScheduleTwoTone",
	"ScissorOutlined",
	"SearchOutlined",
	"SecurityScanFilled",
	"SecurityScanOutlined",
	"SecurityScanTwoTone",
	"SelectOutlined",
	"SendOutlined",
	"SettingFilled",
	"SettingOutlined",
	"SettingTwoTone",
	"ShakeOutlined",
	"ShareAltOutlined",
	"ShopFilled",
	"ShopOutlined",
	"ShopTwoTone",
	"ShoppingCartOutlined",
	"ShoppingFilled",
	"ShoppingOutlined",
	"ShoppingTwoTone",
	"ShrinkOutlined",
	"SignalFilled",
	"SignatureFilled",
	"SignatureOutlined",
	"SisternodeOutlined",
	"SketchCircleFilled",
	"SketchOutlined",
	"SketchSquareFilled",
	"SkinFilled",
	"SkinOutlined",
	"SkinTwoTone",
	"SkypeFilled",
	"SkypeOutlined",
	"SlackCircleFilled",
	"SlackOutlined",
	"SlackSquareFilled",
	"SlackSquareOutlined",
	"SlidersFilled",
	"SlidersOutlined",
	"SlidersTwoTone",
	"SmallDashOutlined",
	"SmileFilled",
	"SmileOutlined",
	"SmileTwoTone",
	"SnippetsFilled",
	"SnippetsOutlined",
	"SnippetsTwoTone",
	"SolutionOutlined",
	"SortAscendingOutlined",
	"SortDescendingOutlined",
	"SoundFilled",
	"SoundOutlined",
	"SoundTwoTone",
	"SplitCellsOutlined",
	"SpotifyFilled",
	"SpotifyOutlined",
	"StarFilled",
	"StarOutlined",
	"StarTwoTone",
	"StepBackwardFilled",
	"StepBackwardOutlined",
	"StepForwardFilled",
	"StepForwardOutlined",
	"StockOutlined",
	"StopFilled",
	"StopOutlined",
	"StopTwoTone",
	"StrikethroughOutlined",
	"SubnodeOutlined",
	"SunFilled",
	"SunOutlined",
	"SwapLeftOutlined",
	"SwapOutlined",
	"SwapRightOutlined",
	"SwitcherFilled",
	"SwitcherOutlined",
	"SwitcherTwoTone",
	"SyncOutlined",
	"TableOutlined",
	"TabletFilled",
	"TabletOutlined",
	"TabletTwoTone",
	"TagFilled",
	"TagOutlined",
	"TagTwoTone",
	"TagsFilled",
	"TagsOutlined",
	"TagsTwoTone",
	"TaobaoCircleFilled",
	"TaobaoCircleOutlined",
	"TaobaoOutlined",
	"TaobaoSquareFilled",
	"TeamOutlined",
	"ThunderboltFilled",
	"ThunderboltOutlined",
	"ThunderboltTwoTone",
	"TikTokFilled",
	"TikTokOutlined",
	"ToTopOutlined",
	"ToolFilled",
	"ToolOutlined",
	"ToolTwoTone",
	"TrademarkCircleFilled",
	"TrademarkCircleOutlined",
	"TrademarkCircleTwoTone",
	"TrademarkOutlined",
	"TransactionOutlined",
	"TranslationOutlined",
	"TrophyFilled",
	"TrophyOutlined",
	"TrophyTwoTone",
	"TruckFilled",
	"TruckOutlined",
	"TwitchFilled",
	"TwitchOutlined",
	"TwitterCircleFilled",
	"TwitterOutlined",
	"TwitterSquareFilled",
	"UnderlineOutlined",
	"UndoOutlined",
	"UngroupOutlined",
	"UnlockFilled",
	"UnlockOutlined",
	"UnlockTwoTone",
	"UnorderedListOutlined",
	"UpCircleFilled",
	"UpCircleOutlined",
	"UpCircleTwoTone",
	"UpOutlined",
	"UpSquareFilled",
	"UpSquareOutlined",
	"UpSquareTwoTone",
	"UploadOutlined",
	"UsbFilled",
	"UsbOutlined",
	"UsbTwoTone",
	"UserAddOutlined",
	"UserDeleteOutlined",
	"UserOutlined",
	"UserSwitchOutlined",
	"UsergroupAddOutlined",
	"UsergroupDeleteOutlined",
	"VerifiedOutlined",
	"VerticalAlignBottomOutlined",
	"VerticalAlignMiddleOutlined",
	"VerticalAlignTopOutlined",
	"VerticalLeftOutlined",
	"VerticalRightOutlined",
	"VideoCameraAddOutlined",
	"VideoCameraFilled",
	"VideoCameraOutlined",
	"VideoCameraTwoTone",
	"WalletFilled",
	"WalletOutlined",
	"WalletTwoTone",
	"WarningFilled",
	"WarningOutlined",
	"WarningTwoTone",
	"WechatFilled",
	"WechatOutlined",
	"WechatWorkFilled",
	"WechatWorkOutlined",
	"WeiboCircleFilled",
	"WeiboCircleOutlined",
	"WeiboOutlined",
	"WeiboSquareFilled",
	"WeiboSquareOutlined",
	"WhatsAppOutlined",
	"WifiOutlined",
	"WindowsFilled",
	"WindowsOutlined",
	"WomanOutlined",
	"XFilled",
	"XOutlined",
	"YahooFilled",
	"YahooOutlined",
	"YoutubeFilled",
	"YoutubeOutlined",
	"YuqueFilled",
	"YuqueOutlined",
	"ZhihuCircleFilled",
	"ZhihuOutlined",
	"ZhihuSquareFilled",
	"ZoomInOutlined",
	"ZoomOutOutlined"
];

//#endregion
//#region src/index.ts
function isExclude(name, exclude) {
	if (!exclude) return false;
	if (typeof exclude === "string") return name === exclude;
	if (exclude instanceof RegExp) return !!name.match(exclude);
	if (Array.isArray(exclude)) {
		for (const item of exclude) if (name === item || name.match(item)) return true;
	}
	return false;
}
/**
* Resolver for [Antdv Next](https://antdv-next.com)
*/
function AntdvNextResolver(options) {
	return {
		type: "component",
		resolve: (name) => {
			const opts = Object.assign({}, options);
			if (isExclude(name, opts.exclude)) return;
			if (opts.resolveIcons && icons_default.includes(name)) return {
				name,
				from: "@antdv-next/icons"
			};
			const importName = components_default[name];
			if (importName) {
				if (isExclude(importName, opts.exclude)) return;
				return {
					name: importName,
					from: "antdv-next"
				};
			}
		}
	};
}

//#endregion
export { AntdvNextResolver };
