// xe-utils@4.0.4/es/static.js
import staticStrUndefined from './staticStrUndefined'
import staticStrLast from './staticStrLast'
import staticStrFirst from './staticStrFirst'
import staticDayTime from './staticDayTime'
import staticWeekTime from './staticWeekTime'
import staticLocation from './staticLocation'
import staticWindow from './staticWindow'
import staticDocument from './staticDocument'
import staticEncodeURIComponent from './staticEncodeURIComponent'
import staticDecodeURIComponent from './staticDecodeURIComponent'
import staticObjectToString from './staticObjectToString'
import staticParseInt from './staticParseInt'
import staticEscapeMap from './staticEscapeMap'
import staticHGKeyRE from './staticHGKeyRE'

var staticExports = {
  staticStrUndefined: staticStrUndefined,
  staticStrLast: staticStrLast,
  staticStrFirst: staticStrFirst,
  staticDayTime: staticDayTime,
  staticWeekTime: staticWeekTime,
  staticLocation: staticLocation,
  staticWindow: staticWindow,
  staticDocument: staticDocument,
  staticEncodeURIComponent: staticEncodeURIComponent,
  staticDecodeURIComponent: staticDecodeURIComponent,
  staticObjectToString: staticObjectToString,
  staticParseInt: staticParseInt,
  staticEscapeMap: staticEscapeMap,
  staticHGKeyRE: staticHGKeyRE
}

export default staticExports
