// @v-c/async-validator@1.0.1/dist/validator/index.js
import any_default from "./any.js";
import array_default from "./array.js";
import boolean_default from "./boolean.js";
import date_default from "./date.js";
import enum_default from "./enum.js";
import float_default from "./float.js";
import integer_default from "./integer.js";
import method_default from "./method.js";
import number_default from "./number.js";
import object_default from "./object.js";
import pattern_default from "./pattern.js";
import regexp_default from "./regexp.js";
import required_default from "./required.js";
import string_default from "./string.js";
import type_default from "./type.js";
var validator_default = {
	string: string_default,
	method: method_default,
	number: number_default,
	boolean: boolean_default,
	regexp: regexp_default,
	integer: integer_default,
	float: float_default,
	array: array_default,
	object: object_default,
	enum: enum_default,
	pattern: pattern_default,
	date: date_default,
	url: type_default,
	hex: type_default,
	email: type_default,
	tel: type_default,
	required: required_default,
	any: any_default
};
export { validator_default as default };
