// antdv-next@1.2.2/dist/timeline/TimelineItem.js
import { defineComponent } from "vue";

//#region src/timeline/TimelineItem.tsx
const TIMELINE_ITEM_MARK = "_ANTDV_NEXT_TIMELINE_ITEM";
const TimelineItem = /* @__PURE__ */ defineComponent(() => {
	return () => null;
}, {
	props: {
		color: { required: false },
		placement: {
			type: String,
			required: false
		},
		position: {
			type: String,
			required: false
		},
		loading: {
			type: Boolean,
			required: false,
			default: void 0
		},
		title: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		content: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		label: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		icon: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		dot: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		}
	},
	name: "ATimelineItem",
	inheritAttrs: false
});
TimelineItem[TIMELINE_ITEM_MARK] = true;

//#endregion
export { TIMELINE_ITEM_MARK, TimelineItem };
