// xe-utils@3.9.1/es/date.js
import getWhatYear from './getWhatYear'
import getWhatQuarter from './getWhatQuarter'
import getWhatMonth from './getWhatMonth'
import getWhatWeek from './getWhatWeek'
import getWhatDay from './getWhatDay'
import getWhatHours from './getWhatHours'
import getWhatMinutes from './getWhatMinutes'
import getWhatSeconds from './getWhatSeconds'
import toStringDate from './toStringDate'
import toDateString from './toDateString'
import now from './now'
import timestamp from './timestamp'
import isValidDate from './isValidDate'
import isDateSame from './isDateSame'
import getYearDay from './getYearDay'
import getYearWeek from './getYearWeek'
import getMonthWeek from './getMonthWeek'
import getDayOfYear from './getDayOfYear'
import getDayOfMonth from './getDayOfMonth'
import getDateDiff from './getDateDiff'

var dateExports = {
  now: now,
  timestamp: timestamp,
  isValidDate: isValidDate,
  isDateSame: isDateSame,
  toStringDate: toStringDate,
  toDateString: toDateString,
  getWhatYear: getWhatYear,
  getWhatQuarter: getWhatQuarter,
  getWhatMonth: getWhatMonth,
  getWhatWeek: getWhatWeek,
  getWhatDay: getWhatDay,
  getWhatHours: getWhatHours,
  getWhatMinutes: getWhatMinutes,
  getWhatSeconds: getWhatSeconds,
  getYearDay: getYearDay,
  getYearWeek: getYearWeek,
  getMonthWeek: getMonthWeek,
  getDayOfYear: getDayOfYear,
  getDayOfMonth: getDayOfMonth,
  getDateDiff: getDateDiff
}

export default dateExports
