// @antdv-next/cssinjs@1.0.6/dist/index.js
import Keyframes_default from "./Keyframes.js";
import autoPrefix_default from "./transformers/autoPrefix.js";
import { StyleProvider, createCache, provideStyleContext, useStyleContext, useStyleContextProvide } from "./StyleContext.js";
import { collectStyleText, setStyleCollector } from "./ssr/styleCollector.js";
import calc_default from "./theme/calc/index.js";
import Theme from "./theme/Theme.js";
import createTheme from "./theme/createTheme.js";
import "./theme/index.js";
import resolveHash_default from "./util/resolveHash.js";
import { token2CSSVar } from "./util/css-variables.js";
import { supportLogicProps, supportWhere, unit } from "./util/index.js";
import useCacheToken, { getComputedToken } from "./hooks/useCacheToken.js";
import legacyNotSelectorLinter_default from "./linters/legacyNotSelectorLinter.js";
import logicalPropertiesLinter_default from "./linters/logicalPropertiesLinter.js";
import NaNLinter_default from "./linters/NaNLinter.js";
import parentSelectorLinter_default from "./linters/parentSelectorLinter.js";
import "./linters/index.js";
import useStyleRegister from "./hooks/useStyleRegister.js";
import useCSSVarRegister from "./hooks/useCSSVarRegister.js";
import extractStyle from "./extractStyle.js";
import legacyLogicalProperties_default from "./transformers/legacyLogicalProperties.js";
import px2rem_default from "./transformers/px2rem.js";
import statistic_default, { merge, statistic } from "./cssinjs-utils/util/statistic.js";
import genStyleUtils_default from "./cssinjs-utils/util/genStyleUtils.js";
import "./cssinjs-utils/index.js";

//#region src/index.ts
const _experimental = { supportModernCSS: () => supportWhere() && supportLogicProps() };

//#endregion
export { Keyframes_default as Keyframes, NaNLinter_default as NaNLinter, StyleProvider, Theme, _experimental, autoPrefix_default as autoPrefixTransformer, collectStyleText, createCache, createTheme, extractStyle, calc_default as genCalc, genStyleUtils_default as genStyleUtils, getComputedToken, resolveHash_default as hash, legacyLogicalProperties_default as legacyLogicalPropertiesTransformer, legacyNotSelectorLinter_default as legacyNotSelectorLinter, logicalPropertiesLinter_default as logicalPropertiesLinter, merge as mergeToken, parentSelectorLinter_default as parentSelectorLinter, provideStyleContext, px2rem_default as px2remTransformer, setStyleCollector, statistic, statistic_default as statisticToken, token2CSSVar, unit, useCSSVarRegister, useCacheToken, useStyleContext, useStyleContextProvide, useStyleRegister };
