// antdv-next@1.2.2/dist/tree/style/index.js
import { genFocusOutline, resetComponent } from "../../style/index.js";
import { genStyleHooks } from "../../theme/util/genStyleUtils.js";
import { mergeToken } from "../../theme/internal.js";
import collapse_default from "../../style/motion/collapse.js";
import { getStyle } from "../../checkbox/style/index.js";
import { genDirectoryStyle } from "./directory.js";
import { Keyframes, unit } from "@antdv-next/cssinjs";

//#region src/tree/style/index.ts
const treeNodeFX = new Keyframes("ant-tree-node-fx-do-not-use", {
	"0%": { opacity: 0 },
	"100%": { opacity: 1 }
});
function getSwitchStyle(prefixCls, token) {
	return { [`.${prefixCls}-switcher-icon`]: {
		display: "inline-block",
		fontSize: 10,
		verticalAlign: "baseline",
		svg: { transition: `transform ${token.motionDurationSlow}` }
	} };
}
function getDropIndicatorStyle(prefixCls, token) {
	return { [`.${prefixCls}-drop-indicator`]: {
		position: "absolute",
		zIndex: 1,
		height: 2,
		backgroundColor: token.colorPrimary,
		borderRadius: 1,
		pointerEvents: "none",
		"&:after": {
			position: "absolute",
			top: -3,
			insetInlineStart: -6,
			width: 8,
			height: 8,
			backgroundColor: "transparent",
			border: `${unit(token.lineWidthBold)} solid ${token.colorPrimary}`,
			borderRadius: "50%",
			content: "\"\""
		}
	} };
}
function genBaseStyle(prefixCls, token) {
	const { treeCls, treeNodeCls, treeNodePadding, titleHeight, indentSize, switcherSize, motionDurationMid, nodeSelectedBg, nodeHoverBg, colorTextQuaternary, controlItemBgActiveDisabled } = token;
	return { [treeCls]: {
		...resetComponent(token),
		["--rc-virtual-list-scrollbar-bg"]: token.colorSplit,
		background: token.colorBgContainer,
		borderRadius: token.borderRadius,
		transition: `background-color ${token.motionDurationSlow}`,
		"&-rtl": { direction: "rtl" },
		[`&${treeCls}-rtl ${treeCls}-switcher_close ${treeCls}-switcher-icon svg`]: { transform: "rotate(90deg)" },
		[`${treeCls}-list`]: { "&:focus-visible": {
			outline: "none",
			[`${treeNodeCls}-active ${treeCls}-node-content-wrapper`]: { ...genFocusOutline(token) }
		} },
		[`${treeCls}-list-holder-inner`]: { alignItems: "flex-start" },
		[`&${treeCls}-block-node`]: { [`${treeCls}-list-holder-inner`]: {
			alignItems: "stretch",
			[`${treeCls}-node-content-wrapper`]: { flex: "auto" },
			[`${treeNodeCls}.dragging:after`]: {
				position: "absolute",
				inset: 0,
				border: `1px solid ${token.colorPrimary}`,
				opacity: 0,
				animationName: treeNodeFX,
				animationDuration: token.motionDurationSlow,
				animationPlayState: "running",
				animationFillMode: "forwards",
				content: "\"\"",
				pointerEvents: "none",
				borderRadius: token.borderRadius
			}
		} },
		[treeNodeCls]: {
			display: "flex",
			alignItems: "flex-start",
			marginBottom: treeNodePadding,
			lineHeight: unit(titleHeight),
			position: "relative",
			"&:before": {
				content: "\"\"",
				position: "absolute",
				zIndex: 1,
				insetInlineStart: 0,
				width: "100%",
				top: "100%",
				height: treeNodePadding
			},
			[`&-disabled ${treeCls}-node-content-wrapper`]: {
				color: token.colorTextDisabled,
				cursor: "not-allowed",
				"&:hover": { background: "transparent" }
			},
			[`${treeCls}-checkbox-disabled + ${treeCls}-node-selected,&${treeNodeCls}-disabled${treeNodeCls}-selected ${treeCls}-node-content-wrapper`]: { backgroundColor: controlItemBgActiveDisabled },
			[`${treeCls}-checkbox-disabled`]: { pointerEvents: "unset" },
			[`&:not(${treeNodeCls}-disabled)`]: { [`${treeCls}-node-content-wrapper`]: { "&:hover": { color: token.nodeHoverColor } } },
			[`&-active ${treeCls}-node-content-wrapper`]: { background: token.controlItemBgHover },
			[`&:not(${treeNodeCls}-disabled).filter-node ${treeCls}-title`]: {
				color: token.colorPrimary,
				fontWeight: token.fontWeightStrong
			},
			"&-draggable": {
				cursor: "grab",
				[`${treeCls}-draggable-icon`]: {
					flexShrink: 0,
					width: switcherSize,
					textAlign: "center",
					visibility: "visible",
					color: colorTextQuaternary
				},
				[`&${treeNodeCls}-disabled ${treeCls}-draggable-icon`]: { visibility: "hidden" }
			}
		},
		[`${treeCls}-indent`]: {
			alignSelf: "stretch",
			whiteSpace: "nowrap",
			userSelect: "none",
			"&-unit": {
				display: "inline-block",
				width: indentSize
			}
		},
		[`${treeCls}-draggable-icon`]: { visibility: "hidden" },
		[`${treeCls}-switcher, ${treeCls}-checkbox`]: { marginInlineEnd: token.calc(token.calc(switcherSize).sub(token.controlInteractiveSize)).div(2).equal() },
		[`${treeCls}-checkbox`]: {
			flexShrink: 0,
			alignSelf: "flex-start",
			marginBlockStart: token.calc(token.calc(titleHeight).sub(token.controlInteractiveSize)).div(2).equal()
		},
		[`${treeCls}-switcher`]: {
			...getSwitchStyle(prefixCls, token),
			position: "relative",
			flex: "none",
			alignSelf: "stretch",
			width: switcherSize,
			textAlign: "center",
			cursor: "pointer",
			userSelect: "none",
			transition: `all ${token.motionDurationSlow}`,
			"&-noop": { cursor: "unset" },
			"&:before": {
				pointerEvents: "none",
				content: "\"\"",
				width: switcherSize,
				height: titleHeight,
				position: "absolute",
				left: {
					_skip_check_: true,
					value: 0
				},
				top: 0,
				borderRadius: token.borderRadius,
				transition: `all ${token.motionDurationSlow}`
			},
			[`&:not(${treeCls}-switcher-noop):hover:before`]: { backgroundColor: token.colorBgTextHover },
			[`&_close ${treeCls}-switcher-icon svg`]: { transform: "rotate(-90deg)" },
			"&-loading-icon": { color: token.colorPrimary },
			"&-leaf-line": {
				position: "relative",
				zIndex: 1,
				display: "inline-block",
				width: "100%",
				height: "100%",
				"&:before": {
					position: "absolute",
					top: 0,
					insetInlineEnd: token.calc(switcherSize).div(2).equal(),
					bottom: token.calc(treeNodePadding).mul(-1).equal(),
					marginInlineStart: -1,
					borderInlineEnd: `1px solid ${token.colorBorder}`,
					content: "\"\""
				},
				"&:after": {
					position: "absolute",
					width: token.calc(token.calc(switcherSize).div(2).equal()).mul(.8).equal(),
					height: token.calc(titleHeight).div(2).equal(),
					borderBottom: `1px solid ${token.colorBorder}`,
					content: "\"\""
				}
			}
		},
		[`${treeCls}-node-content-wrapper`]: {
			position: "relative",
			minHeight: titleHeight,
			paddingBlock: 0,
			paddingInline: token.paddingXS,
			background: "transparent",
			borderRadius: token.borderRadius,
			cursor: "pointer",
			transition: [
				`all ${motionDurationMid}`,
				"border 0s",
				"line-height 0s",
				"box-shadow 0s"
			].join(", "),
			...getDropIndicatorStyle(prefixCls, token),
			"&:hover": { backgroundColor: nodeHoverBg },
			[`&${treeCls}-node-selected`]: {
				color: token.nodeSelectedColor,
				backgroundColor: nodeSelectedBg
			},
			[`${treeCls}-iconEle`]: {
				display: "inline-block",
				width: switcherSize,
				height: titleHeight,
				textAlign: "center",
				verticalAlign: "top",
				"&:empty": { display: "none" }
			}
		},
		[`${treeCls}-unselectable ${treeCls}-node-content-wrapper:hover`]: { backgroundColor: "transparent" },
		[`${treeNodeCls}.drop-container > [draggable]`]: { boxShadow: `0 0 0 2px ${token.colorPrimary}` },
		"&-show-line": {
			[`${treeCls}-indent-unit`]: {
				position: "relative",
				height: "100%",
				"&:before": {
					position: "absolute",
					top: 0,
					insetInlineEnd: token.calc(switcherSize).div(2).equal(),
					bottom: token.calc(treeNodePadding).mul(-1).equal(),
					borderInlineEnd: `1px solid ${token.colorBorder}`,
					content: "\"\""
				},
				"&-end:before": { display: "none" }
			},
			[`${treeCls}-switcher`]: {
				background: "transparent",
				"&-line-icon": { verticalAlign: "-0.15em" }
			}
		},
		[`${treeNodeCls}-leaf-last ${treeCls}-switcher-leaf-line:before`]: {
			top: "auto !important",
			bottom: "auto !important",
			height: `${unit(token.calc(titleHeight).div(2).equal())} !important`
		}
	} };
}
function genTreeStyle(prefixCls, token, enableDirectory = true) {
	const treeCls = `.${prefixCls}`;
	const treeToken = mergeToken(token, {
		treeCls,
		treeNodeCls: `${treeCls}-treenode`,
		treeNodePadding: token.calc(token.paddingXS).div(2).equal()
	});
	return [genBaseStyle(prefixCls, treeToken), enableDirectory && genDirectoryStyle(treeToken)].filter(Boolean);
}
function initComponentToken(token) {
	const { controlHeightSM, controlItemBgHover, controlItemBgActive } = token;
	const titleHeight = controlHeightSM;
	return {
		titleHeight,
		switcherSize: titleHeight,
		indentSize: titleHeight,
		nodeHoverBg: controlItemBgHover,
		nodeHoverColor: token.colorText,
		nodeSelectedBg: controlItemBgActive,
		nodeSelectedColor: token.colorText
	};
}
const prepareComponentToken = (token) => {
	const { colorTextLightSolid, colorPrimary } = token;
	return {
		...initComponentToken(token),
		directoryNodeSelectedColor: colorTextLightSolid,
		directoryNodeSelectedBg: colorPrimary
	};
};
var style_default = genStyleHooks("Tree", (token, { prefixCls }) => [
	{ [token.componentCls]: getStyle(`${prefixCls}-checkbox`, token) },
	genTreeStyle(prefixCls, token),
	collapse_default(token)
], prepareComponentToken);

//#endregion
export { style_default as default, genBaseStyle, genTreeStyle, initComponentToken, prepareComponentToken };
