// antdv-next@1.2.2/dist/notification/PurePanel.js
import { useBaseConfig, useComponentBaseConfig } from "../config-provider/context.js";
import { pureAttrs, useMergeSemantic, useToArr, useToProps } from "../_util/hooks/useMergeSemantic.js";
import { getSlotPropsFnRun, toPropsRefs } from "../_util/tools.js";
import useCSSVarCls_default from "../config-provider/hooks/useCSSVarCls.js";
import useClosable, { pickClosable } from "../_util/hooks/useClosable.js";
import style_default from "./style/index.js";
import pure_panel_default from "./style/pure-panel.js";
import { computed, createVNode, defineComponent, mergeDefaults, mergeProps } from "vue";
import { clsx } from "@v-c/util";
import { CheckCircleFilled, CloseCircleFilled, CloseOutlined, ExclamationCircleFilled, InfoCircleFilled } from "@antdv-next/icons";
import { omit } from "es-toolkit";
import { Notice } from "@v-c/notification";

//#region src/notification/PurePanel.tsx
function getCloseIcon(prefixCls, closeIcon) {
	if (closeIcon === null || closeIcon === false) return null;
	return closeIcon || createVNode(CloseOutlined, { "class": `${prefixCls}-close-icon` }, null);
}
const typeToIcon = {
	success: CheckCircleFilled,
	info: InfoCircleFilled,
	error: CloseCircleFilled,
	warning: ExclamationCircleFilled
};
const defaults = { role: "alert" };
const PureContent = /* @__PURE__ */ defineComponent((props) => {
	return () => {
		const { prefixCls, icon, type, title, description, actions, role = "alert", styles, classes: pureContentCls } = props;
		let iconNode;
		if (icon) iconNode = createVNode("span", {
			"class": clsx(`${prefixCls}-icon`, pureContentCls.icon),
			"style": styles.icon
		}, [icon]);
		else if (type) if (typeToIcon[type]) iconNode = createVNode(typeToIcon[type], {
			class: clsx(`${prefixCls}-icon`, pureContentCls.icon, `${prefixCls}-icon-${type}`),
			style: styles.icon
		});
		else iconNode = null;
		return createVNode("div", {
			"class": clsx({ [`${prefixCls}-with-icon`]: iconNode }),
			"role": role
		}, [
			iconNode,
			title && createVNode("div", {
				"class": clsx(`${prefixCls}-title`, pureContentCls.title),
				"style": styles.title
			}, [title]),
			description && createVNode("div", {
				"class": clsx(`${prefixCls}-description`, pureContentCls.description),
				"style": styles.description
			}, [description]),
			actions && createVNode("div", {
				"class": clsx(`${prefixCls}-actions`, pureContentCls.actions),
				"style": styles.actions
			}, [actions])
		]);
	};
}, {
	props: /* @__PURE__ */ mergeDefaults({
		prefixCls: {
			type: String,
			required: true
		},
		icon: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		title: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		description: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		actions: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		type: {
			type: String,
			required: false
		},
		role: {
			type: String,
			required: false
		},
		classes: {
			type: Object,
			required: true
		},
		styles: {
			type: Object,
			required: true
		}
	}, defaults),
	name: "NoticePureContent",
	inheritAttrs: false
});
const omitKeys = [
	"prefixCls",
	"icon",
	"type",
	"message",
	"title",
	"description",
	"btn",
	"actions",
	"closeIcon",
	"className",
	"style",
	"styles",
	"classNames",
	"closable"
];
/** @private Internal Component. Do not use in your production. */
const PurePanel = /* @__PURE__ */ defineComponent((props, { attrs, slots }) => {
	const { classes: notificationClassNames, styles } = toPropsRefs(props, "classes", "styles");
	const { getPrefixCls, class: contextClassName, style: contextStyle, classes: contextClassNames, styles: contextStyles } = useComponentBaseConfig("notification", props);
	const { notification: notificationContext } = useBaseConfig("notification", props);
	const prefixCls = computed(() => props.prefixCls ?? getPrefixCls("notification"));
	const mergedProps = computed(() => props);
	const rootCls = useCSSVarCls_default(prefixCls);
	const [hashId, cssVarCls] = style_default(prefixCls, rootCls);
	const [mergedClassNames, mergedStyles] = useMergeSemantic(useToArr(contextClassNames, notificationClassNames), useToArr(contextStyles, styles), useToProps(mergedProps));
	const closeableInfo = useClosable(pickClosable(computed(() => props)), pickClosable(notificationContext), computed(() => {
		return {
			closable: true,
			closeIcon: createVNode(CloseOutlined, { "class": `${prefixCls.value}-close-icon` }, null),
			closeIconRender: (icon) => getCloseIcon(prefixCls.value, icon)
		};
	}));
	const rawClosable = computed(() => closeableInfo.value?.[0]);
	const mergedCloseIcon = computed(() => closeableInfo.value?.[1]);
	const ariaProps = computed(() => closeableInfo.value?.[3] || {});
	const mergedClosable = computed(() => {
		const { closable } = props;
		return rawClosable.value ? {
			onClose: closable && typeof closable === "object" ? closable?.onClose : void 0,
			closeIcon: mergedCloseIcon.value,
			...ariaProps.value
		} : false;
	});
	return () => {
		const noticePrefixCls = `${prefixCls.value}-notice`;
		const notificationClassName = attrs.class;
		const style = attrs.style;
		const restProps = omit(props, omitKeys);
		const actions = getSlotPropsFnRun(slots, props, "actions");
		return createVNode("div", {
			"class": clsx(`${noticePrefixCls}-pure-panel`, hashId.value, notificationClassName, cssVarCls.value, rootCls.value, mergedClassNames.value?.root),
			"style": mergedStyles.value.root
		}, [createVNode(pure_panel_default, { "prefixCls": prefixCls.value }, null), createVNode(Notice, mergeProps({ "style": {
			...contextStyle.value,
			...style
		} }, pureAttrs(attrs), restProps, {
			"prefixCls": prefixCls.value,
			"eventKey": "pure",
			"duration": null,
			"closable": mergedClosable.value,
			"class": clsx(notificationClassName, contextClassName.value),
			"content": createVNode(PureContent, {
				"classes": mergedClassNames.value,
				"styles": mergedStyles.value,
				"prefixCls": noticePrefixCls,
				"icon": props.icon,
				"type": props.type,
				"title": props.title,
				"description": props.description,
				"actions": actions
			}, null)
		}), null)]);
	};
}, {
	props: {
		content: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		duration: {
			type: [
				Number,
				Boolean,
				null
			],
			required: false,
			default: void 0
		},
		showProgress: {
			type: Boolean,
			required: false,
			default: void 0
		},
		pauseOnHover: {
			type: Boolean,
			required: false,
			default: void 0
		},
		closable: {
			type: [Boolean, Object],
			required: false,
			default: void 0
		},
		prefixCls: {
			type: String,
			required: false
		},
		classes: {
			type: [Object, Function],
			required: false
		},
		styles: {
			type: [Object, Function],
			required: false
		},
		closeIcon: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		props: {
			type: Object,
			required: false
		},
		onClose: { required: false },
		onClick: {
			type: Function,
			required: false
		},
		icon: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		title: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		description: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		actions: {
			type: [
				Function,
				String,
				Number,
				null,
				Object,
				Boolean
			],
			required: false,
			default: void 0
		},
		type: {
			type: String,
			required: false
		},
		role: {
			type: String,
			required: false
		}
	},
	name: "NoticePurePanel",
	inheritAttrs: false
});
var PurePanel_default = PurePanel;

//#endregion
export { PureContent, PurePanel_default as default, getCloseIcon };
